--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/preload/Preload.java
@@ -1,0 +_,26 @@
+package one.tranic.sewlia.preload;
+
+import one.tranic.sewlia.reflect.Reflect;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Set;
+
+public class Preload {
+    private final Logger logger = LoggerFactory.getLogger("Preload");
+
+    public Preload() {
+        try {
+            @Nullable Set<Preloads> set = Reflect.findAllClass("one.tranic.sewlia.preload.mod", Preloads.class);
+            if (set != null && !set.isEmpty()) {
+                for (Preloads preload : set) {
+                    preload.doPreload();
+                }
+            }
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+        }
+    }
+}
