From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 10 Feb 2025 13:51:44 +0800
Subject: [PATCH] Sewlia VT


diff --git a/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java b/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java
index 28857d0c9b53f2068d51b8f09ef40df7a2b97502..b46133e6a37a9547e31a1b2a14946d4504086698 100644
--- a/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java
+++ b/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java
@@ -429,6 +429,7 @@ public final class PluginRemapper {
             TimeUnit.SECONDS,
             ScalingThreadPool.createUnboundedQueue(),
             new ThreadFactoryBuilder()
+                .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.paper.PluginRemapper.getThreadFactory())
                 .setNameFormat("Paper Plugin Remapper Thread - %1$d")
                 .setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER))
                 .build(),
diff --git a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
index 374abffb9f1ce1a308822aed13038e77fe9ca08b..77be3946a5561c77cf5e2f38868fc800b6afc3e2 100644
--- a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
@@ -31,7 +31,7 @@ public final class FoliaAsyncScheduler implements AsyncScheduler {
 
             @Override
             public Thread newThread(final Runnable run) {
-                final Thread ret = new Thread(run);
+                final Thread ret = one.tranic.sewlia.config.mod.feature.virtual_thread.folia.AsyncScheduler.getThread(run);
 
                 ret.setName("Folia Async Scheduler Thread #" + this.idGenerator.getAndIncrement());
                 ret.setPriority(Thread.NORM_PRIORITY - 1);
@@ -47,7 +47,7 @@ public final class FoliaAsyncScheduler implements AsyncScheduler {
     private final ScheduledExecutorService timerThread = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
         @Override
         public Thread newThread(final Runnable run) {
-            final Thread ret = new Thread(run);
+            final Thread ret = one.tranic.sewlia.config.mod.feature.virtual_thread.folia.AsyncScheduler.getThread(run);
 
             ret.setName("Folia Async Scheduler Thread Timer");
             ret.setPriority(Thread.NORM_PRIORITY + 1);
diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 4dcec640f5870d713bd3b98389a45dbef8a4ea8a..4a9844d825934137b8b9c2f7e17cd51a555539f6 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -38,6 +38,7 @@ public final class MCUtil {
         }
     };
     public static final ExecutorService ASYNC_EXECUTOR = Executors.newFixedThreadPool(2, new ThreadFactoryBuilder()
+        .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.paper.AsyncTaskHandler.getThreadFactory())
         .setNameFormat("Paper Async Task Handler Thread - %1$d")
         .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(MinecraftServer.LOGGER))
         .build()
diff --git a/src/main/java/io/papermc/paper/util/ServerWorkerThreadFactory.java b/src/main/java/io/papermc/paper/util/ServerWorkerThreadFactory.java
new file mode 100644
index 0000000000000000000000000000000000000000..e9a33a28555a3d478fa45d49aeca167acd63f1fc
--- /dev/null
+++ b/src/main/java/io/papermc/paper/util/ServerWorkerThreadFactory.java
@@ -0,0 +1,30 @@
+package io.papermc.paper.util;
+
+import net.minecraft.Util;
+import one.tranic.sewlia.config.mod.feature.virtual_thread.paper.Worker;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ServerWorkerThreadFactory implements ThreadFactory {
+    private static final AtomicInteger threadId = new AtomicInteger(1);
+    private final String poolName;
+    private final int prioritityModifier;
+
+    public ServerWorkerThreadFactory(String poolName, int prioritityModifier) {
+        this.poolName = poolName;
+        this.prioritityModifier = prioritityModifier;
+    }
+
+    @Override
+    public Thread newThread(@NotNull Runnable r) {
+        final Thread ret = Worker.value ? Thread.ofVirtual().unstarted(r) : new Thread(r);
+        ret.setName("Worker-" + poolName + "-" + threadId.getAndIncrement());
+        ret.setPriority(Thread.NORM_PRIORITY + prioritityModifier); // Deprioritize over main
+        ret.setDaemon(true);
+        ret.setUncaughtExceptionHandler(Util::onThreadException);
+
+        return ret;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java b/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
index 8a4f95049c63afb28bef6719c77b7a7092e75aae..3a7ba39c0bf59443392eacae547c7a58ed14f653 100644
--- a/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
@@ -53,7 +53,7 @@ public class Main {
             URLClassLoader classLoader = new URLClassLoader(extractedUrls.toArray(new URL[0]));
 
             System.out.println("Starting server");
-            Thread runThread = new Thread(() -> {
+            Thread runThread = one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit.ServerMain.getThread(() -> {
                 try {
                     Class<?> mainClass = Class.forName(mainClassName, true, classLoader);
                     MethodHandle mainHandle = MethodHandles.lookup().findStatic(mainClass, "main", MethodType.methodType(void.class, String[].class)).asFixedArity();
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
index 0ca279fb71d39c81b1f608e0ee9ba3e498d55fa3..c5a67dd94641794e6e59aa2d2c9339c1f03378af 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
@@ -40,9 +40,13 @@ public class CraftAsyncScheduler extends CraftScheduler {
 
     private final ThreadPoolExecutor executor = new ThreadPoolExecutor(
             4, Integer.MAX_VALUE,30L, TimeUnit.SECONDS, new SynchronousQueue<>(),
-            new ThreadFactoryBuilder().setNameFormat("Craft Scheduler Thread - %1$d").build());
+            new ThreadFactoryBuilder().setNameFormat("Craft Scheduler Thread - %1$d")
+                    .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit.CraftScheduler.getThreadFactory())
+                    .build());
     private final Executor management = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
-            .setNameFormat("Craft Async Scheduler Management Thread").build());
+            .setNameFormat("Craft Async Scheduler Management Thread")
+            .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit.AsyncScheduler.getThreadFactory())
+            .build());
     private final List<CraftTask> temp = new ArrayList<>();
 
     CraftAsyncScheduler() {
