From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 22 Feb 2025 00:34:34 +0800
Subject: [PATCH] Modify CraftMetaItem


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index 7d85c554e4e5df49d4965dbed9c112b86cd70727..1f55f4852738c10ae3010221656558d11057f35e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -41,6 +41,9 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+
+import it.unimi.dsi.fastutil.objects.ObjectArraySet;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
@@ -314,12 +317,12 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
     private Integer damage; // Paper - may not be set
     private Integer maxDamage;
 
-    private static final Set<DataComponentType> HANDLED_TAGS = Sets.newHashSet();
+    private static final Set<DataComponentType> HANDLED_TAGS = new ObjectOpenHashSet<>();
     private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
 
     private CompoundTag customTag;
     protected DataComponentPatch.Builder unhandledTags = DataComponentPatch.builder();
-    private Set<DataComponentType<?>> removedTags = Sets.newHashSet();
+    private Set<DataComponentType<?>> removedTags = new ObjectOpenHashSet<>();
     private CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(CraftMetaItem.DATA_TYPE_REGISTRY);
 
     private int version = CraftMagicNumbers.INSTANCE.getDataVersion(); // Internal use only
@@ -333,7 +336,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
         this.itemName = meta.itemName;
 
         if (meta.lore != null) {
-            this.lore = new ArrayList<Component>(meta.lore);
+            this.lore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(meta.lore);
         }
 
         if (meta.hasCustomModelData()) {
@@ -409,7 +412,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
         CraftMetaItem.getOrEmpty(tag, CraftMetaItem.LORE).ifPresent((l) -> {
             List<Component> list = l.lines();
-            this.lore = new ArrayList<Component>(list.size());
+            this.lore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(list.size());
             for (int index = 0; index < list.size(); index++) {
                 Component line = list.get(index);
                 this.lore.add(line);
@@ -619,7 +622,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
         Iterable<?> lore = SerializableMeta.getObject(Iterable.class, map, CraftMetaItem.LORE.BUKKIT, true);
         if (lore != null) {
-            CraftMetaItem.safelyAdd(lore, this.lore = new ArrayList<Component>(), true);
+            CraftMetaItem.safelyAdd(lore, this.lore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(), true);
         }
 
         Object customModelData = SerializableMeta.getObject(Object.class, map, CraftMetaItem.CUSTOM_MODEL_DATA.BUKKIT, true);
@@ -635,7 +638,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
         Object blockData = SerializableMeta.getObject(Object.class, map, CraftMetaItem.BLOCK_DATA.BUKKIT, true);
         if (blockData != null) {
-            Map<String, String> mapBlockData = new HashMap<>();
+            Map<String, String> mapBlockData = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
 
             if (blockData instanceof Map) {
                 for (Entry<?, ?> entry : ((Map<?, ?>) blockData).entrySet()) {
@@ -1372,13 +1375,13 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
     @Override
     public List<String> getLore() {
-        return this.lore == null ? null : new ArrayList<String>(Lists.transform(this.lore, CraftChatMessage::fromComponent));
+        return this.lore == null ? null : new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(Lists.transform(this.lore, CraftChatMessage::fromComponent));
     }
 
     // Paper start
     @Override
     public List<net.md_5.bungee.api.chat.BaseComponent[]> getLoreComponents() {
-        return this.lore == null ? null : new ArrayList<>(this.lore.stream().map(entry ->
+        return this.lore == null ? null : new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(this.lore.stream().map(entry ->
             net.md_5.bungee.chat.ComponentSerializer.parse(CraftChatMessage.toJSON(entry))
         ).collect(java.util.stream.Collectors.toList()));
     }
@@ -1390,7 +1393,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             this.lore = null;
         } else {
             if (this.lore == null) {
-                this.lore = new ArrayList<Component>(lore.size());
+                this.lore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(lore.size());
             } else {
                 this.lore.clear();
             }
@@ -1406,7 +1409,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             this.lore = null;
         } else {
             if (this.lore == null) {
-                safelyAdd(lore, this.lore = new ArrayList<>(lore.size()), false);
+                safelyAdd(lore, this.lore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(lore.size()), false);
             } else {
                 this.lore.clear();
                 safelyAdd(lore, this.lore, false);
@@ -1866,7 +1869,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
             if (componentValue.isPresent()) {
                 net.minecraft.nbt.Tag componentValueAsNBT = (net.minecraft.nbt.Tag) ((DataComponentType) componentType).codecOrThrow().encodeStart(ops, componentValue.get()).getOrThrow();
-                String componentValueAsNBTString = new SnbtPrinterTagVisitor("", 0, new ArrayList<>()).visit(componentValueAsNBT);
+                String componentValueAsNBTString = new SnbtPrinterTagVisitor("", 0, new it.unimi.dsi.fastutil.objects.ObjectArrayList<>()).visit(componentValueAsNBT);
                 componentString.add(componentKey + "=" + componentValueAsNBTString);
             } else {
                 componentString.add("!" + componentKey);
@@ -2068,7 +2071,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
         try {
             CraftMetaItem clone = (CraftMetaItem) super.clone();
             if (this.lore != null) {
-                clone.lore = new ArrayList<Component>(this.lore);
+                clone.lore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(this.lore);
             }
             if (this.hasCustomModelData()) {
                 clone.customModelData = new CraftCustomModelDataComponent(this.customModelData);
@@ -2084,7 +2087,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             if (this.customTag != null) {
                 clone.customTag = this.customTag.copy();
             }
-            clone.removedTags = Sets.newHashSet(this.removedTags);
+            clone.removedTags = new ObjectOpenHashSet<>(this.removedTags);
             clone.persistentDataContainer = new CraftPersistentDataContainer(this.persistentDataContainer.getTagsCloned(), CraftMetaItem.DATA_TYPE_REGISTRY); // Paper - deep clone NBT tags
             clone.hideFlag = this.hideFlag;
             clone.hideTooltip = this.hideTooltip;
@@ -2151,7 +2154,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
         if (this.hasLore()) {
             // SPIGOT-7625: Convert lore to json before serializing it
-            List<String> jsonLore = new ArrayList<>();
+            List<String> jsonLore = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
 
             for (Component component : this.lore) {
                 jsonLore.add(CraftChatMessage.toJSON(component));
@@ -2177,7 +2180,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             builder.put(CraftMetaItem.REPAIR.BUKKIT, this.repairCost);
         }
 
-        List<String> hideFlags = new ArrayList<String>();
+        List<String> hideFlags = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         for (ItemFlag hideFlagEnum : this.getItemFlags()) {
             hideFlags.add(CraftItemFlag.bukkitToString(hideFlagEnum));
         }
@@ -2253,7 +2256,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             builder.put(CraftMetaItem.MAX_DAMAGE.BUKKIT, this.maxDamage);
         }
 
-        final Map<String, net.minecraft.nbt.Tag> internalTags = new HashMap<String, net.minecraft.nbt.Tag>();
+        final Map<String, net.minecraft.nbt.Tag> internalTags = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
         this.serializeInternal(internalTags);
         if (!internalTags.isEmpty()) {
             CompoundTag internal = new CompoundTag();
@@ -2302,7 +2305,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             RegistryAccess registryAccess = CraftRegistry.getMinecraftRegistry();
             Registry<DataComponentType<?>> componentTypeRegistry = registryAccess.lookupOrThrow(Registries.DATA_COMPONENT_TYPE);
 
-            List<String> removedTags = new ArrayList<>();
+            List<String> removedTags = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
             for (DataComponentType<?> removed : this.removedTags) {
                 String componentKey = componentTypeRegistry.getResourceKey(removed).orElseThrow().location().toString();
 
@@ -2350,7 +2353,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             return;
         }
 
-        Map<String, List<Object>> mods = new LinkedHashMap<>();
+        Map<String, List<Object>> mods = new it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap<>();
         for (Map.Entry<Attribute, AttributeModifier> entry : modifiers.entries()) {
             if (entry.getKey() == null) {
                 continue;
@@ -2359,7 +2362,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             if (modCollection == null || modCollection.isEmpty()) {
                 continue;
             }
-            mods.put(CraftAttribute.bukkitToString(entry.getKey()), new ArrayList<>(modCollection));
+            mods.put(CraftAttribute.bukkitToString(entry.getKey()), new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(modCollection));
         }
         builder.put(key.BUKKIT, mods);
     }
@@ -2427,8 +2430,8 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
 
     // Paper start - improve checking handled tags
     @org.jetbrains.annotations.VisibleForTesting
-    public static final Map<Class<? extends CraftMetaItem>, Set<DataComponentType<?>>> HANDLED_DCTS_PER_TYPE = new HashMap<>();
-    protected static final Set<DataComponentType<?>> DEFAULT_HANDLED_DCTS = Set.of(
+    public static final Map<Class<? extends CraftMetaItem>, Set<DataComponentType<?>>> HANDLED_DCTS_PER_TYPE = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
+    protected static final Set<DataComponentType<?>> DEFAULT_HANDLED_DCTS = ObjectArraySet.of(
         CraftMetaItem.NAME.TYPE,
         CraftMetaItem.ITEM_NAME.TYPE,
         CraftMetaItem.LORE.TYPE,
@@ -2463,36 +2466,36 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
     public static Set<DataComponentType<?>> getTopLevelHandledDcts(final Class<? extends CraftMetaItem> clazz) {
         synchronized (HANDLED_DCTS_PER_TYPE) {
             if (HANDLED_DCTS_PER_TYPE.isEmpty()) {
-                final Map<Class<? extends CraftMetaItem>, Set<DataComponentType<?>>> map = new HashMap<>();
-                map.put(CraftMetaArmor.class, Set.of(CraftMetaArmor.TRIM.TYPE));
-                map.put(CraftMetaArmorStand.class, Set.of(CraftMetaArmorStand.ENTITY_TAG.TYPE));
-                map.put(CraftMetaAxolotlBucket.class, Set.of(CraftMetaAxolotlBucket.ENTITY_TAG.TYPE, CraftMetaAxolotlBucket.BUCKET_ENTITY_TAG.TYPE));
-                map.put(CraftMetaBanner.class, Set.of(CraftMetaBanner.PATTERNS.TYPE)); // banner uses same tag as block state
-                map.put(CraftMetaShield.class, Set.of(CraftMetaShield.BASE_COLOR.TYPE, CraftMetaBanner.PATTERNS.TYPE));
-                map.put(CraftMetaBlockState.class, Set.of(CraftMetaBlockState.BLOCK_ENTITY_TAG.TYPE));
-                map.put(CraftMetaBook.class, Set.of(CraftMetaBook.BOOK_CONTENT.TYPE));
-                map.put(CraftMetaBookSigned.class, Set.of(CraftMetaBookSigned.BOOK_CONTENT.TYPE));
-                map.put(CraftMetaBundle.class, Set.of(CraftMetaBundle.ITEMS.TYPE));
-                map.put(CraftMetaCharge.class, Set.of(CraftMetaCharge.EXPLOSION.TYPE));
-                map.put(CraftMetaColorableArmor.class, Set.of(CraftMetaArmor.TRIM.TYPE, CraftMetaLeatherArmor.COLOR.TYPE));
-                map.put(CraftMetaCompass.class, Set.of(CraftMetaCompass.LODESTONE_TARGET.TYPE));
-                map.put(CraftMetaCrossbow.class, Set.of(CraftMetaCrossbow.CHARGED_PROJECTILES.TYPE));
-                map.put(CraftMetaEnchantedBook.class, Set.of(CraftMetaEnchantedBook.STORED_ENCHANTMENTS.TYPE));
-                map.put(CraftMetaEntityTag.class, Set.of(CraftMetaEntityTag.ENTITY_TAG.TYPE));
-                map.put(CraftMetaFirework.class, Set.of(CraftMetaFirework.FIREWORKS.TYPE));
-                map.put(CraftMetaKnowledgeBook.class, Set.of(CraftMetaKnowledgeBook.BOOK_RECIPES.TYPE));
-                map.put(CraftMetaLeatherArmor.class, Set.of(CraftMetaLeatherArmor.COLOR.TYPE));
-                map.put(CraftMetaMap.class, Set.of(CraftMetaMap.MAP_COLOR.TYPE, CraftMetaMap.MAP_POST_PROCESSING.TYPE, CraftMetaMap.MAP_ID.TYPE));
-                map.put(CraftMetaMusicInstrument.class, Set.of(CraftMetaMusicInstrument.GOAT_HORN_INSTRUMENT.TYPE));
-                map.put(CraftMetaOminousBottle.class, Set.of(CraftMetaOminousBottle.OMINOUS_BOTTLE_AMPLIFIER.TYPE));
-                map.put(CraftMetaPotion.class, Set.of(CraftMetaPotion.POTION_CONTENTS.TYPE));
-                map.put(CraftMetaSkull.class, Set.of(CraftMetaSkull.SKULL_PROFILE.TYPE, CraftMetaSkull.NOTE_BLOCK_SOUND.TYPE));
-                map.put(CraftMetaSpawnEgg.class, Set.of(CraftMetaSpawnEgg.ENTITY_TAG.TYPE));
-                map.put(CraftMetaSuspiciousStew.class, Set.of(CraftMetaSuspiciousStew.EFFECTS.TYPE));
-                map.put(CraftMetaTropicalFishBucket.class, Set.of(CraftMetaTropicalFishBucket.ENTITY_TAG.TYPE, CraftMetaTropicalFishBucket.BUCKET_ENTITY_TAG.TYPE));
+                final Map<Class<? extends CraftMetaItem>, Set<DataComponentType<?>>> map = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
+                map.put(CraftMetaArmor.class, ObjectArraySet.of(CraftMetaArmor.TRIM.TYPE));
+                map.put(CraftMetaArmorStand.class, ObjectArraySet.of(CraftMetaArmorStand.ENTITY_TAG.TYPE));
+                map.put(CraftMetaAxolotlBucket.class, ObjectArraySet.of(CraftMetaAxolotlBucket.ENTITY_TAG.TYPE, CraftMetaAxolotlBucket.BUCKET_ENTITY_TAG.TYPE));
+                map.put(CraftMetaBanner.class, ObjectArraySet.of(CraftMetaBanner.PATTERNS.TYPE)); // banner uses same tag as block state
+                map.put(CraftMetaShield.class, ObjectArraySet.of(CraftMetaShield.BASE_COLOR.TYPE, CraftMetaBanner.PATTERNS.TYPE));
+                map.put(CraftMetaBlockState.class, ObjectArraySet.of(CraftMetaBlockState.BLOCK_ENTITY_TAG.TYPE));
+                map.put(CraftMetaBook.class, ObjectArraySet.of(CraftMetaBook.BOOK_CONTENT.TYPE));
+                map.put(CraftMetaBookSigned.class, ObjectArraySet.of(CraftMetaBookSigned.BOOK_CONTENT.TYPE));
+                map.put(CraftMetaBundle.class, ObjectArraySet.of(CraftMetaBundle.ITEMS.TYPE));
+                map.put(CraftMetaCharge.class, ObjectArraySet.of(CraftMetaCharge.EXPLOSION.TYPE));
+                map.put(CraftMetaColorableArmor.class, ObjectArraySet.of(CraftMetaArmor.TRIM.TYPE, CraftMetaLeatherArmor.COLOR.TYPE));
+                map.put(CraftMetaCompass.class, ObjectArraySet.of(CraftMetaCompass.LODESTONE_TARGET.TYPE));
+                map.put(CraftMetaCrossbow.class, ObjectArraySet.of(CraftMetaCrossbow.CHARGED_PROJECTILES.TYPE));
+                map.put(CraftMetaEnchantedBook.class, ObjectArraySet.of(CraftMetaEnchantedBook.STORED_ENCHANTMENTS.TYPE));
+                map.put(CraftMetaEntityTag.class, ObjectArraySet.of(CraftMetaEntityTag.ENTITY_TAG.TYPE));
+                map.put(CraftMetaFirework.class, ObjectArraySet.of(CraftMetaFirework.FIREWORKS.TYPE));
+                map.put(CraftMetaKnowledgeBook.class, ObjectArraySet.of(CraftMetaKnowledgeBook.BOOK_RECIPES.TYPE));
+                map.put(CraftMetaLeatherArmor.class, ObjectArraySet.of(CraftMetaLeatherArmor.COLOR.TYPE));
+                map.put(CraftMetaMap.class, ObjectArraySet.of(CraftMetaMap.MAP_COLOR.TYPE, CraftMetaMap.MAP_POST_PROCESSING.TYPE, CraftMetaMap.MAP_ID.TYPE));
+                map.put(CraftMetaMusicInstrument.class, ObjectArraySet.of(CraftMetaMusicInstrument.GOAT_HORN_INSTRUMENT.TYPE));
+                map.put(CraftMetaOminousBottle.class, ObjectArraySet.of(CraftMetaOminousBottle.OMINOUS_BOTTLE_AMPLIFIER.TYPE));
+                map.put(CraftMetaPotion.class, ObjectArraySet.of(CraftMetaPotion.POTION_CONTENTS.TYPE));
+                map.put(CraftMetaSkull.class, ObjectArraySet.of(CraftMetaSkull.SKULL_PROFILE.TYPE, CraftMetaSkull.NOTE_BLOCK_SOUND.TYPE));
+                map.put(CraftMetaSpawnEgg.class, ObjectArraySet.of(CraftMetaSpawnEgg.ENTITY_TAG.TYPE));
+                map.put(CraftMetaSuspiciousStew.class, ObjectArraySet.of(CraftMetaSuspiciousStew.EFFECTS.TYPE));
+                map.put(CraftMetaTropicalFishBucket.class, ObjectArraySet.of(CraftMetaTropicalFishBucket.ENTITY_TAG.TYPE, CraftMetaTropicalFishBucket.BUCKET_ENTITY_TAG.TYPE));
 
                 for (final Map.Entry<Class<? extends CraftMetaItem>, Set<DataComponentType<?>>> entry : map.entrySet()) {
-                    final ArrayList<DataComponentType<?>> topLevelTags = new ArrayList<>(entry.getValue());
+                    final List<DataComponentType<?>> topLevelTags = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(entry.getValue());
                     // add tags common to CraftMetaItem to all
                     topLevelTags.addAll(DEFAULT_HANDLED_DCTS);
                     HANDLED_DCTS_PER_TYPE.put(entry.getKey(), Set.copyOf(topLevelTags));
@@ -2593,7 +2596,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
     }
 
     private static List<net.minecraft.advancements.critereon.BlockPredicate> convertFromLegacyNamespaced(final Collection<com.destroystokyo.paper.Namespaced> namespaceds) {
-        final List<net.minecraft.advancements.critereon.BlockPredicate> predicates = new ArrayList<>();
+        final List<net.minecraft.advancements.critereon.BlockPredicate> predicates = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         final net.minecraft.core.Registry<net.minecraft.world.level.block.Block> blockRegistry = net.minecraft.server.MinecraftServer.getServer().registryAccess().lookupOrThrow(net.minecraft.core.registries.Registries.BLOCK);
         for (final com.destroystokyo.paper.Namespaced namespaced : namespaceds) {
             if (namespaced instanceof final org.bukkit.NamespacedKey key) {
@@ -2606,7 +2609,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
     }
 
     private static Set<com.destroystokyo.paper.Namespaced> convertToLegacyNamespaced(final Collection<net.minecraft.advancements.critereon.BlockPredicate> predicates) {
-        final Set<com.destroystokyo.paper.Namespaced> namespaceds = Sets.newHashSet();
+        final Set<com.destroystokyo.paper.Namespaced> namespaceds = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
         for (final net.minecraft.advancements.critereon.BlockPredicate predicate : predicates) {
             if (predicate.blocks().isEmpty()) {
                 continue;
