From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 11 Feb 2025 14:21:53 +0800
Subject: [PATCH] Adjust gridExponent to avoid additional errors


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index de88c3d9d3523a7bd3f3dcbfc62d72658192521d..3f139f16559f53045fad60ce0e2b83f89bc6bc5f 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -403,7 +403,7 @@ public class GlobalConfiguration extends ConfigurationPart {
     public class ThreadedRegions extends ConfigurationPart {
 
         public int threads = -1;
-        public int gridExponent = 4;
+        public int gridExponent = 2;
 
         @PostProcess
         public void postProcess() {
