--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -616,6 +_,7 @@
             DefaultPermissions.registerCorePermissions();
             CraftDefaultPermissions.registerCorePermissions();
             if (!io.papermc.paper.configuration.GlobalConfiguration.get().misc.loadPermissionsYmlBeforePlugins) this.loadCustomPermissions(); // Paper
+            one.tranic.sewlia.permission.Permission.registerCorePermissions(); // Sewlia
             this.helpMap.initializeCommands();
             this.syncCommands();
         }
@@ -3225,6 +_,7 @@
         if (!io.papermc.paper.configuration.GlobalConfiguration.get().misc.loadPermissionsYmlBeforePlugins) loadCustomPermissions();
         DefaultPermissions.registerCorePermissions();
         CraftDefaultPermissions.registerCorePermissions();
+        one.tranic.sewlia.permission.Permission.registerCorePermissions(); // Sewlia
     }
 
     @Override
