--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1367,7 +_,8 @@
                 player.setRealHealth(health);
             }
 
-            player.updateScaledHealth(false);
+            //player.updateScaledHealth(false); // Leaf - Commented out to fix sprint glitch
+            this.entityData.set(LivingEntity.DATA_HEALTH_ID, player.getScaledHealth()); // Leaf - Sprint glitch fixed by pafias
             return;
         }
         // CraftBukkit end
