--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java
@@ -1,0 +_,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class UserRepository {
+    @DisableReload
+    @Comment("Users add a central repository mirror")
+    public static String value = "";
+}
