--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/Enable.java
@@ -1,0 +_,11 @@
+package one.tranic.sewlia.config.mod.misc.cache_player_profile;
+
+import one.tranic.sewlia.config.annotation.Comments;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Enable {
+    @DisableReload
+    @Comments({"Cache the player profile result on they first join.",
+            "It's useful if Mojang's verification server is down."})
+    public static boolean value = false;
+}
