--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/ResultTimeout.java
@@ -1,0 +_,10 @@
+package one.tranic.sewlia.config.mod.misc.cache_player_profile;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ResultTimeout {
+    @DisableReload
+    @Comment("The timeout of the cache. Unit: Minutes.")
+    public static int value = 1440;
+}
