--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -212,6 +_,8 @@
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
+        one.tranic.sewlia.command.CommandUtil.register(); // Sewlia - commands
+        one.tranic.su.util.RunTask.checkThreadTask(); // Sewlia - Thread Check
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
@@ -349,6 +_,8 @@
                 MinecraftServerStatistics.registerJmxMonitoring(this);
                 LOGGER.info("JMX monitoring enabled");
             }
+
+            one.tranic.sewlia.task.BossBarTask.startAll(); // Sewlia
 
             return true;
         }
