From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 11 Feb 2025 14:21:53 +0800
Subject: [PATCH] Adjust gridExponent to avoid additional errors


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index ee6748a212c51f163dda29617e591a38c0257ae2..1f7d84c4690415a215b7bc2b03439d924a9d87ef 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -408,7 +408,7 @@ public class GlobalConfiguration extends ConfigurationPart {
     public class ThreadedRegions extends ConfigurationPart {
 
         public int threads = -1;
-        public int gridExponent = 4;
+        public int gridExponent = 2;
 
         @PostProcess
         public void postProcess() {
