From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 22 Feb 2025 12:37:45 +0800
Subject: [PATCH] Modify CraftParticle


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftParticle.java b/src/main/java/org/bukkit/craftbukkit/CraftParticle.java
index f9e33ec8ba7adaf729fecabc093aece201accf8f..5f0e261f17c8584a70b5169a597f19a58767e8d7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftParticle.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftParticle.java
@@ -88,12 +88,12 @@ public abstract class CraftParticle<D> implements Keyed {
 
     private final NamespacedKey key;
     private final net.minecraft.core.particles.ParticleType<?> particle;
-    private final Class<D> clazz;
+    //private final Class<D> clazz;
 
     public CraftParticle(NamespacedKey key, net.minecraft.core.particles.ParticleType<?> particle, Class<D> clazz) {
         this.key = key;
         this.particle = particle;
-        this.clazz = clazz;
+        //this.clazz = clazz;
     }
 
     public net.minecraft.core.particles.ParticleType<?> getHandle() {
@@ -109,7 +109,7 @@ public abstract class CraftParticle<D> implements Keyed {
 
     public static class CraftParticleRegistry extends CraftRegistry<CraftParticle<?>, net.minecraft.core.particles.ParticleType<?>> {
 
-        private static final Map<NamespacedKey, BiFunction<NamespacedKey, net.minecraft.core.particles.ParticleType<?>, CraftParticle<?>>> PARTICLE_MAP = new HashMap<>();
+        private static final Map<NamespacedKey, BiFunction<NamespacedKey, net.minecraft.core.particles.ParticleType<?>, CraftParticle<?>>> PARTICLE_MAP = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
 
         private static final BiFunction<NamespacedKey, net.minecraft.core.particles.ParticleType<?>, CraftParticle<?>> VOID_FUNCTION = (name, particle) -> new CraftParticle<>(name, particle, Void.class) {
             @Override
