From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 10 Feb 2025 15:50:47 +0800
Subject: [PATCH] KioCG: Chunk API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index de8b9048c8395c05b8688bc9d984b8ad680f15b3..f42692cd4f0154705c3d5b030d281cfc333803ed 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -437,4 +437,12 @@ public class CraftChunk implements Chunk {
     static {
         Arrays.fill(FULL_LIGHT, (byte) 0xFF);
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getChunkHotAvg() {
+        final net.minecraft.world.level.chunk.LevelChunk target = this.worldServer.getChunkIfLoaded(this.x,this.z);
+        return target == null ? -1 : target.getChunkHot().getAverage();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 714504664aaed8b6a7a3e46b2b18816263d622cf..1382a5a77aefecb45c092fe513af60dbe9ff3e7c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3618,4 +3618,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void setDeathScreenScore(final int score) {
         getHandle().setScore(score);
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getNearbyChunkHot() {
+        return this.getHandle().getNearbyChunkHot();
+    }
+    // KioCG end
 }
