--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Gui.java
@@ -1,0 +_,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Gui {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable r) {
+        return value ? Thread.ofVirtual().unstarted(r) : new Thread(r);
+    }
+}
