From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 28 Oct 2024 15:39:36 +0800
Subject: [PATCH] Sewlia Command


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 78f33298e809a7f6d079d9f2c64e2caa47a1b25a..a88addb2d75f5f27627ee1a557956d994922b521 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -223,6 +223,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
+        one.tranic.sewlia.command.CommandUtil.register(); // Sewlia - commands
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
diff --git a/src/main/java/one/tranic/sewlia/command/CommandUtil.java b/src/main/java/one/tranic/sewlia/command/CommandUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..d53dcdf04784a31b5571f420f6b2e6eea88faf83
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/command/CommandUtil.java
@@ -0,0 +1,23 @@
+package one.tranic.sewlia.command;
+
+import net.intelie.tinymap.TinySet;
+import one.tranic.sewlia.Reflect;
+import org.bukkit.Bukkit;
+import org.bukkit.command.Command;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+
+public class CommandUtil {
+    public static void register() {
+        try {
+            @Nullable TinySet<Command> set = Reflect.findAllClass("one.tranic.sewlia.command.mod", Command.class);
+            if (set == null || set.isEmpty()) return;
+            for (Command clazz : set) {
+                Bukkit.getCommandMap().register(clazz.getName(), "minecraft", clazz);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}
