From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 9 Nov 2024 12:34:47 +0800
Subject: [PATCH] Lithium: cached_hashcode

Source in: https://github.com/CaffeineMC/lithium-fabric

diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index 9b952e74ee705256cbe9390553432f6364d4f2c1..7aac9e1677ffa45997bd85b0f4c990acf94e75ce 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -579,7 +579,18 @@ public class Block extends BlockBehaviour implements ItemLike {
     }
     // Spigot end
 
-    private static record ShapePairKey(VoxelShape first, VoxelShape second) {
+    private static final class ShapePairKey {
+
+        private final VoxelShape first;
+        private final VoxelShape second;
+        private final int hash; // Lithium: cached_hashcode
+
+        public ShapePairKey(VoxelShape first, VoxelShape second) {
+            this.first = first;
+            this.second = second;
+
+            this.hash = System.identityHashCode(this.first) * 31 + System.identityHashCode(this.second); // Lithium - cached_hashcode
+        }
 
         public boolean equals(Object object) {
             boolean flag;
@@ -596,7 +607,7 @@ public class Block extends BlockBehaviour implements ItemLike {
         }
 
         public int hashCode() {
-            return System.identityHashCode(this.first) * 31 + System.identityHashCode(this.second);
+            return this.hash; // Lithium: cached_hashcode
         }
     }
 }
