--- a/net/minecraft/world/entity/projectile/LargeFireball.java
+++ b/net/minecraft/world/entity/projectile/LargeFireball.java
@@ -71,4 +_,11 @@
             this.bukkitYield = this.explosionPower = compound.getByte("ExplosionPower");
         }
     }
+
+    // Sewlia start
+    @Override
+    public boolean shouldBeSaved() {
+        return !one.tranic.sewlia.config.mod.misc.dont_save.LargeFireball.value;
+    }
+    // Sewlia end
 }
