--- a/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -132,4 +_,11 @@
         super.readAdditionalSaveData(tag);
         this.setDangerous(tag.getBoolean("dangerous"));
     }
+
+    // Sewlia start
+    @Override
+    public boolean shouldBeSaved() {
+        return !one.tranic.sewlia.config.mod.misc.dont_save.WitherSkull.value;
+    }
+    // Sewlia end
 }
