--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/ConfigUtils.java
@@ -1,0 +_,57 @@
+package one.tranic.sewlia.config;
+
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+
+public class ConfigUtils {
+    private static YamlConfiguration configuration;
+
+    public static YamlConfiguration getConfiguration() {
+        return configuration;
+    }
+
+    public static void reloadConfiguration(boolean isCommandSource) {
+        try {
+            File configFile = new File("sewlia.yaml");
+            if (!configFile.exists()) {
+                configFile.createNewFile();
+            }
+            configuration = YamlConfiguration.loadConfiguration(configFile);
+            if (!isCommandSource) addDefaults(configFile);
+            readAll(isCommandSource);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void addDefaults(@NotNull File file) throws IOException {
+        configuration.options().setHeader(
+                Arrays.asList(
+                        "Sewlia. Configuration",
+                        "Sewlia Github: https://github.com/LevelTranic/Sewlia",
+                        "It is recommended to always keep the latest version, download it at https://tranic.one/downloads/sewlia"
+                )
+        );
+
+        @Nullable Map<Class<?>, String> clz = ConfigScanner.getClasss();
+        if (clz != null && !clz.isEmpty()) {
+            for (Map.Entry<Class<?>, String> entry : clz.entrySet())
+                ConfigScanner.processStaticValueFieldWithWrite(entry.getKey(), entry.getValue());
+        }
+        configuration.options().copyDefaults(true);
+        configuration.save(file);
+    }
+
+    private static void readAll(boolean isReload) {
+        @Nullable Map<Class<?>, String> clz = ConfigScanner.getClasss();
+        if (clz == null || clz.isEmpty()) return;
+        for (Map.Entry<Class<?>, String> entry : clz.entrySet())
+            ConfigScanner.processStaticValueFieldWithRead(entry.getKey(), entry.getValue(), isReload);
+    }
+}
