From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 10 Feb 2025 21:26:28 +0800
Subject: [PATCH] enderportal


diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 982c79f83dc9e469ab55deecdf9cfe3f076cdb75..5ff8735f59055ae98d5407108bedb20b16360c2d 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -401,7 +401,7 @@ public class FallingBlockEntity extends Entity {
         ResourceKey<Level> resourceKey1 = this.level().dimension();
         boolean flag = (resourceKey1 == Level.END || resourceKey == Level.END) && resourceKey1 != resourceKey;
         Entity entity = super.teleport(teleportTransition);
-        this.forceTickAfterTeleportToDuplicate = entity != null && flag && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowUnsafeEndPortalTeleportation; // Paper
+        this.forceTickAfterTeleportToDuplicate = entity != null && flag; // Paper // Sewlia
         return entity;
     }
 
diff --git a/net/minecraft/world/level/block/EndPortalBlock.java b/net/minecraft/world/level/block/EndPortalBlock.java
index 177735cf744e564081e4c140a0f8210c3a07e037..5ff5cf15360f1586b4fb3be711a8dfeef3a56ded 100644
--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -67,6 +67,7 @@ public class EndPortalBlock extends BaseEntityBlock implements Portal {
                 if (level.paperConfig().misc.disableEndCredits) {serverPlayer.seenCredits = true; return;} // Paper - Option to disable end credits
                 serverPlayer.showEndCredits();
             } else {
+                if (!(entity instanceof net.minecraft.world.entity.player.Player)) entity.endPortalLogicAsync(pos); // Sewlia
                 entity.setAsInsidePortal(this, pos);
             }
         }
