From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 10 Feb 2025 16:15:05 +0800
Subject: [PATCH] Show Region Player Location


diff --git a/io/papermc/paper/threadedregions/TickRegions.java b/io/papermc/paper/threadedregions/TickRegions.java
index 988fe74578065c9464f5639e5cc6af79619edef5..27ea9ac0500f4a219d262ba5ce78f9a907d1268c 100644
--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -7,13 +7,16 @@ import com.mojang.logging.LogUtils;
 import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.longs.Long2ReferenceMap;
 import it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.Reference2ReferenceMap;
 import it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import org.slf4j.Logger;
 import java.util.Iterator;
+import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
@@ -134,6 +137,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
         private final AtomicInteger entityCount = new AtomicInteger();
         private final AtomicInteger playerCount = new AtomicInteger();
         private final AtomicInteger chunkCount = new AtomicInteger();
+        private final ConcurrentLinkedDeque<ServerPlayer> players = new ConcurrentLinkedDeque<>();
 
         public int getEntityCount() {
             return this.entityCount.get();
@@ -147,10 +151,16 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
             return this.chunkCount.get();
         }
 
+        public ObjectArrayList<ServerPlayer> getPlayers() {
+            return new ObjectArrayList<>(this.players);
+        }
+
         void updateFrom(final RegionizedWorldData data) {
             this.entityCount.setRelease(data == null ? 0 : data.getEntityCount());
             this.playerCount.setRelease(data == null ? 0 : data.getPlayerCount());
             this.chunkCount.setRelease(data == null ? 0 : data.getChunkCount());
+            this.players.clear();
+            if (data != null) this.players.addAll(data.getLocalPlayers());
         }
 
         static void updateCurrentRegion() {
diff --git a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
index 5e6b490ee58a90fd7c02fa09093830c0d9c67f6b..85465f7d23ef6857325a1e263b258ad6d59b20ac 100644
--- a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
+++ b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
@@ -1,12 +1,12 @@
 package io.papermc.paper.threadedregions.commands;
 
 import io.papermc.paper.threadedregions.RegionizedServer;
-import io.papermc.paper.threadedregions.RegionizedWorldData;
 import io.papermc.paper.threadedregions.ThreadedRegionizer;
 import io.papermc.paper.threadedregions.TickData;
 import io.papermc.paper.threadedregions.TickRegionScheduler;
 import io.papermc.paper.threadedregions.TickRegions;
 import it.unimi.dsi.fastutil.doubles.DoubleArrayList;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectObjectImmutablePair;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.TextComponent;
@@ -16,6 +16,7 @@ import net.kyori.adventure.text.format.NamedTextColor;
 import net.kyori.adventure.text.format.TextColor;
 import net.kyori.adventure.text.format.TextDecoration;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.level.ChunkPos;
 import org.bukkit.Bukkit;
 import org.bukkit.World;
@@ -24,6 +25,8 @@ import org.bukkit.command.CommandSender;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.entity.Entity;
 import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+
 import java.text.DecimalFormat;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -68,18 +71,51 @@ public final class CommandServerHealth extends Command {
             .build();
     }
 
-    private static Component formatRegionStats(final TickRegions.RegionStats stats, final boolean newline) {
-        return Component.text()
-            .append(Component.text("Chunks: ", PRIMARY))
-            .append(Component.text(NO_DECIMAL_PLACES.get().format((long)stats.getChunkCount()), INFORMATION))
-            .append(Component.text(" Players: ", PRIMARY))
-            .append(Component.text(NO_DECIMAL_PLACES.get().format((long)stats.getPlayerCount()), INFORMATION))
-            .append(Component.text(" Entities: ", PRIMARY))
-            .append(Component.text(NO_DECIMAL_PLACES.get().format((long)stats.getEntityCount()) + (newline ? "\n" : ""), INFORMATION))
-            .build();
+    // Sewlia start - Show Region Player Location
+    private static Component formatRegionStats(final TickRegions.RegionStats stats, final boolean newline, final boolean hasPermission) {
+        TextComponent.Builder builder = Component.text();
+        builder.append(Component.text("Chunks: ", PRIMARY));
+        builder.append(Component.text(NO_DECIMAL_PLACES.get().format(stats.getChunkCount()), INFORMATION));
+
+        TextComponent.Builder playerBuild = Component.text();
+        if (hasPermission) {
+            ObjectArrayList<ServerPlayer> players = stats.getPlayers();
+            final TextComponent component = getTextComponent(players);
+            playerBuild.append(Component.text(" Players: ", PRIMARY).hoverEvent(HoverEvent.showText(component)));
+        } else {
+            playerBuild.append(Component.text(" Players: ", PRIMARY));
+        }
+        builder.append(playerBuild.build());
+        builder.append(Component.text(NO_DECIMAL_PLACES.get().format(stats.getPlayerCount()), INFORMATION));
+        builder.append(Component.text(" Entities: ", PRIMARY));
+        builder.append(Component.text(NO_DECIMAL_PLACES.get().format(stats.getEntityCount()) + (newline ? "\n" : ""), INFORMATION));
+        return builder.build();
     }
 
-    private static boolean executeRegion(final CommandSender sender, final String commandLabel, final String[] args) {
+    private static @NotNull TextComponent getTextComponent(ObjectArrayList<ServerPlayer> players) {
+        TextComponent.@NotNull Builder builder = Component.text();
+        builder.append(Component.text("Region Player List: ", NamedTextColor.GREEN));
+        int size = players.size();
+        for (int i = 0; i < size; i++) {
+            if (size <= 15 || i <= 15) {
+                ServerPlayer player = players.get(i);
+                builder.append(Component.text("\n")).
+                        append(Component.text(player.gameProfile.getName(), NamedTextColor.YELLOW))
+                        .append(Component.text(" | In ", NamedTextColor.WHITE))
+                        .append(Component.text(player.getOnPos().getX() + ", " + player.getOnPos().getY() + ", " + player.getOnPos().getZ(), NamedTextColor.AQUA));
+            } else {
+                builder.append(Component.text("\n"))
+                        .append(Component.text("And the remaining ", NamedTextColor.YELLOW))
+                        .append(Component.text(size - 15, NamedTextColor.AQUA))
+                        .append(Component.text(" players", NamedTextColor.YELLOW));
+                break;
+            }
+        }
+        return builder.build();
+    }
+    // Sewlia end
+
+    private static boolean executeRegion(final CommandSender sender) {
         final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
             TickRegionScheduler.getCurrentRegion();
         if (region == null) {
@@ -120,7 +156,7 @@ public final class CommandServerHealth extends Command {
                 formatRegionInfo("1m: ", util1m, mspt1m, tps1m, true)
             )
             .append(
-                formatRegionStats(region.getData().getRegionStats(), false)
+                    formatRegionStats(region.getData().getRegionStats(), false, (sender instanceof Player && sender.hasPermission("sewlia.feature.tpslocation"))) // Sewlia - Show Region Player Location
             )
 
             .build();
@@ -130,7 +166,7 @@ public final class CommandServerHealth extends Command {
         return true;
     }
 
-    private static boolean executeServer(final CommandSender sender, final String commandLabel, final String[] args) {
+    private static boolean executeServer(final CommandSender sender, final String[] args) {
         final int lowestRegionsCount;
         if (args.length < 2) {
             lowestRegionsCount = 3;
@@ -144,7 +180,7 @@ public final class CommandServerHealth extends Command {
         }
 
         final List<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>> regions =
-            new ArrayList<>();
+            new ObjectArrayList<>();
 
         for (final World bukkitWorld : Bukkit.getWorlds()) {
             final ServerLevel world = ((CraftWorld)bukkitWorld).getHandle();
@@ -157,7 +193,7 @@ public final class CommandServerHealth extends Command {
         double totalUtil = 0.0;
 
         final DoubleArrayList tpsByRegion = new DoubleArrayList();
-        final List<TickData.TickReportData> reportsByRegion = new ArrayList<>();
+        final List<TickData.TickReportData> reportsByRegion = new ObjectArrayList<>();
         final int maxThreadCount = TickRegions.getScheduler().getTotalThreadCount();
 
         final long currTime = System.nanoTime();
@@ -194,7 +230,7 @@ public final class CommandServerHealth extends Command {
         }
 
         final List<ObjectObjectImmutablePair<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, TickData.TickReportData>>
-            regionsBelowThreshold = new ArrayList<>();
+            regionsBelowThreshold = new ObjectArrayList<>();
 
         for (int i = 0, len = regions.size(); i < len; ++i) {
             final TickData.TickReportData report = reportsByRegion.get(i);
@@ -214,7 +250,7 @@ public final class CommandServerHealth extends Command {
 
         final TextComponent.Builder lowestRegionsBuilder = Component.text();
 
-        if (sender instanceof Player) {
+        if (sender instanceof Player && sender.hasPermission("minecraft.command.execute")) {
             lowestRegionsBuilder.append(Component.text(" Click to teleport\n", SECONDARY));
         }
         for (int i = 0, len = Math.min(lowestRegionsCount, regionsBelowThreshold.size()); i < len; ++i) {
@@ -244,26 +280,32 @@ public final class CommandServerHealth extends Command {
 
             final int yLoc = 80;
             final String location = "[w:'" + world.getWorld().getName() + "'," + centerBlockX + "," + yLoc + "," + centerBlockZ + "]";
-            final Component line = Component.text()
-                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
-                .append(Component.text("Region around block ", PRIMARY))
-                .append(Component.text(location, INFORMATION))
-                .append(Component.text(":\n", PRIMARY))
-
-                .append(Component.text("    ", PRIMARY))
-                .append(Component.text(ONE_DECIMAL_PLACES.get().format(util * 100.0), CommandUtil.getUtilisationColourRegion(util)))
-                .append(Component.text("% util at ", PRIMARY))
-                .append(Component.text(TWO_DECIMAL_PLACES.get().format(mspt), CommandUtil.getColourForMSPT(mspt)))
-                .append(Component.text(" MSPT at ", PRIMARY))
-                .append(Component.text(TWO_DECIMAL_PLACES.get().format(tps), CommandUtil.getColourForTPS(tps)))
-                .append(Component.text(" TPS\n", PRIMARY))
-
-                .append(Component.text("    ", PRIMARY))
-                .append(formatRegionStats(region.getData().getRegionStats(), (i + 1) != len))
-                .build()
 
-                .clickEvent(ClickEvent.clickEvent(ClickEvent.Action.RUN_COMMAND, "/minecraft:execute as @s in " + world.getWorld().getKey().toString() + " run tp " + centerBlockX + ".5 " + yLoc + " " + centerBlockZ + ".5"))
-                .hoverEvent(HoverEvent.hoverEvent(HoverEvent.Action.SHOW_TEXT, Component.text("Click to teleport to " + location, SECONDARY)));
+            // Sewlia start
+            TextComponent.@NotNull Builder builder = Component.text();
+            builder.append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                    .append(Component.text("Region around block ", PRIMARY))
+                    .append(Component.text(location, INFORMATION))
+                    .append(Component.text(":\n", PRIMARY))
+                    .append(Component.text("    ", PRIMARY))
+                    .append(Component.text(ONE_DECIMAL_PLACES.get().format(util * 100.0), CommandUtil.getUtilisationColourRegion(util)))
+                    .append(Component.text("% util at ", PRIMARY))
+                    .append(Component.text(TWO_DECIMAL_PLACES.get().format(mspt), CommandUtil.getColourForMSPT(mspt)))
+                    .append(Component.text(" MSPT at ", PRIMARY))
+                    .append(Component.text(TWO_DECIMAL_PLACES.get().format(tps), CommandUtil.getColourForTPS(tps)))
+                    .append(Component.text(" TPS\n", PRIMARY))
+                    .append(Component.text("    ", PRIMARY))
+                    .append(formatRegionStats(region.getData().getRegionStats(), (i + 1) != len, (sender instanceof Player && sender.hasPermission("sewlia.feature.tpslocation")))); // Sewlia - Show Region Player Location
+
+            final Component line;
+            if (sender instanceof Player && sender.hasPermission("minecraft.command.execute")) {
+                line = builder.build()
+                        .clickEvent(ClickEvent.clickEvent(ClickEvent.Action.RUN_COMMAND, "/minecraft:execute as @s in " + world.getWorld().getKey() + " run tp " + centerBlockX + ".5 " + yLoc + " " + centerBlockZ + ".5"))
+                        .hoverEvent(HoverEvent.hoverEvent(HoverEvent.Action.SHOW_TEXT, Component.text("Click to teleport to " + location, SECONDARY)));
+            } else {
+                line = builder.build();
+            }
+            // Sewlia end
 
             lowestRegionsBuilder.append(line);
         }
@@ -318,7 +360,7 @@ public final class CommandServerHealth extends Command {
     }
 
     @Override
-    public boolean execute(final CommandSender sender, final String commandLabel, final String[] args) {
+    public boolean execute(final @NotNull CommandSender sender, final @NotNull String commandLabel, final String[] args) {
         final String type;
         if (args.length < 1) {
             type = "server";
@@ -326,37 +368,37 @@ public final class CommandServerHealth extends Command {
             type = args[0];
         }
 
-        switch (type.toLowerCase(Locale.ROOT)) {
-            case "server": {
-                return executeServer(sender, commandLabel, args);
-            }
-            case "region": {
+        // Sewlia start
+        return switch (type.toLowerCase(Locale.ROOT)) {
+            case "server" -> executeServer(sender, args);
+            case "region" -> {
                 if (!(sender instanceof Entity)) {
                     sender.sendMessage(Component.text("Cannot see current region information as console", NamedTextColor.RED));
-                    return true;
+                    yield true;
                 }
-                return executeRegion(sender, commandLabel, args);
+                yield executeRegion(sender);
             }
-            default: {
+            default -> {
                 sender.sendMessage(Component.text("Type '" + args[0] + "' must be one of: [server, region]", NamedTextColor.RED));
-                return true;
+                yield true;
             }
-        }
+        };
+        // Sewlia end
     }
 
     @Override
-    public List<String> tabComplete(final CommandSender sender, final String alias, final String[] args) throws IllegalArgumentException {
+    public @NotNull List<String> tabComplete(final @NotNull CommandSender sender, final @NotNull String alias, final String[] args) throws IllegalArgumentException {
         if (args.length == 0) {
             if (sender instanceof Entity) {
                 return CommandUtil.getSortedList(Arrays.asList("server", "region"));
             } else {
-                return CommandUtil.getSortedList(Arrays.asList("server"));
+                return CommandUtil.getSortedList(List.of("server"));
             }
         } else if (args.length == 1) {
             if (sender instanceof Entity) {
                 return CommandUtil.getSortedList(Arrays.asList("server", "region"), args[0]);
             } else {
-                return CommandUtil.getSortedList(Arrays.asList("server"), args[0]);
+                return CommandUtil.getSortedList(List.of("server"), args[0]);
             }
         }
         return new ArrayList<>();
diff --git a/io/papermc/paper/threadedregions/commands/CommandUtil.java b/io/papermc/paper/threadedregions/commands/CommandUtil.java
index 1054e28e54f55e0a70eb25aee89cbb4898446fa5..f8e9909926b04462f511b5db85ddd322fad2bda5 100644
--- a/io/papermc/paper/threadedregions/commands/CommandUtil.java
+++ b/io/papermc/paper/threadedregions/commands/CommandUtil.java
@@ -1,5 +1,6 @@
 package io.papermc.paper.threadedregions.commands;
 
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import net.kyori.adventure.text.format.NamedTextColor;
 import net.kyori.adventure.text.format.TextColor;
 import net.kyori.adventure.util.HSVLike;
@@ -12,7 +13,7 @@ import java.util.function.Function;
 public final class CommandUtil {
 
     public static List<String> getSortedList(final Iterable<String> iterable) {
-        final List<String> ret = new ArrayList<>();
+        final List<String> ret = new ObjectArrayList<>();
         for (final String val : iterable) {
             ret.add(val);
         }
@@ -23,7 +24,7 @@ public final class CommandUtil {
     }
 
     public static List<String> getSortedList(final Iterable<String> iterable, final String prefix) {
-        final List<String> ret = new ArrayList<>();
+        final List<String> ret = new ObjectArrayList<>();
         for (final String val : iterable) {
             if (val.regionMatches(0, prefix, 0, prefix.length())) {
                 ret.add(val);
@@ -36,7 +37,7 @@ public final class CommandUtil {
     }
 
     public static <T> List<String> getSortedList(final Iterable<T> iterable, final Function<T, String> transform) {
-        final List<String> ret = new ArrayList<>();
+        final List<String> ret = new ObjectArrayList<>();
         for (final T val : iterable) {
             final String transformed = transform.apply(val);
             if (transformed != null) {
@@ -50,7 +51,7 @@ public final class CommandUtil {
     }
 
     public static <T> List<String> getSortedList(final Iterable<T> iterable, final Function<T, String> transform, final String prefix) {
-        final List<String> ret = new ArrayList<>();
+        final List<String> ret = new ObjectArrayList<>();
         for (final T val : iterable) {
             final String string = transform.apply(val);
             if (string != null && string.regionMatches(0, prefix, 0, prefix.length())) {
