From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 21 Feb 2025 23:22:35 +0800
Subject: [PATCH] Modify Nbt


diff --git a/net/minecraft/nbt/CompoundTag.java b/net/minecraft/nbt/CompoundTag.java
index 361bc458e0bb590c43da60a1cd993a2785ee45e9..9dff28b0a60d8c51f9210e162810354e8158bba3 100644
--- a/net/minecraft/nbt/CompoundTag.java
+++ b/net/minecraft/nbt/CompoundTag.java
@@ -492,7 +492,7 @@ public class CompoundTag implements Tag {
     }
 
     protected CompoundTag shallowCopy() {
-        return new CompoundTag(new HashMap<>(this.tags));
+        return new CompoundTag(new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(this.tags));
     }
 
     @Override
diff --git a/net/minecraft/nbt/ListTag.java b/net/minecraft/nbt/ListTag.java
index 8bbe48a4d7d1771d17bec60b70bdf8e086bfd1c7..3ddd2274fabaf96d012ca64ec82d23e07ecbc0fc 100644
--- a/net/minecraft/nbt/ListTag.java
+++ b/net/minecraft/nbt/ListTag.java
@@ -35,7 +35,7 @@ public class ListTag extends CollectionTag<Tag> {
             } else {
                 accounter.accountBytes(4L, _int);
                 TagType<?> type = TagTypes.getType(_byte);
-                List<Tag> list = Lists.newArrayListWithCapacity(_int);
+                List<Tag> list = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(_int);
 
                 for (int i = 0; i < _int; i++) {
                     list.add(type.load(input, accounter));
@@ -142,7 +142,7 @@ public class ListTag extends CollectionTag<Tag> {
     }
 
     public ListTag() {
-        this(Lists.newArrayList(), (byte)0);
+        this(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(), (byte)0);
     }
 
     @Override
@@ -363,8 +363,8 @@ public class ListTag extends CollectionTag<Tag> {
 
     @Override
     public ListTag copy() {
-        Iterable<Tag> iterable = (Iterable<Tag>)(TagTypes.getType(this.type).isValue() ? this.list : Iterables.transform(this.list, Tag::copy));
-        List<Tag> list = Lists.newArrayList(iterable);
+        Iterable<Tag> iterable = TagTypes.getType(this.type).isValue() ? this.list : Iterables.transform(this.list, Tag::copy);
+        List<Tag> list = one.tranic.t.utils.Collections.newArrayList(iterable);
         return new ListTag(list, this.type);
     }
 
diff --git a/net/minecraft/nbt/NbtOps.java b/net/minecraft/nbt/NbtOps.java
index 7c67bbebdb1a82f71eefbc51cb2049ad5c291d92..1e1d61ce1b9491d50d3ccc8d23bec33bcf44cc88 100644
--- a/net/minecraft/nbt/NbtOps.java
+++ b/net/minecraft/nbt/NbtOps.java
@@ -144,7 +144,7 @@ public class NbtOps implements DynamicOps<Tag> {
             return DataResult.error(() -> "mergeToMap called with not a map: " + map, map);
         } else {
             CompoundTag compoundTag1 = map instanceof CompoundTag compoundTag ? compoundTag.shallowCopy() : new CompoundTag();
-            List<Tag> list = new ArrayList<>();
+            List<Tag> list = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
             otherMap.entries().forEach(pair -> {
                 Tag tag = pair.getFirst();
                 if (!(tag instanceof StringTag)) {
@@ -163,7 +163,7 @@ public class NbtOps implements DynamicOps<Tag> {
             return DataResult.error(() -> "mergeToMap called with not a map: " + tag, tag);
         } else {
             CompoundTag compoundTag1 = tag instanceof CompoundTag compoundTag ? compoundTag.shallowCopy() : new CompoundTag();
-            List<Tag> list = new ArrayList<>();
+            List<Tag> list = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
 
             for (Entry<Tag, Tag> entry : map.entrySet()) {
                 Tag tag1 = entry.getKey();
diff --git a/net/minecraft/nbt/SnbtPrinterTagVisitor.java b/net/minecraft/nbt/SnbtPrinterTagVisitor.java
index 6e3f118fe124c2a9d8ef57c0f3cd25c7323c6df4..e5181f7514ec2107195f1f3217b1c5cb6eb6e834 100644
--- a/net/minecraft/nbt/SnbtPrinterTagVisitor.java
+++ b/net/minecraft/nbt/SnbtPrinterTagVisitor.java
@@ -14,10 +14,10 @@ import java.util.regex.Pattern;
 import net.minecraft.Util;
 
 public class SnbtPrinterTagVisitor implements TagVisitor {
-    private static final Map<String, List<String>> KEY_ORDER = Util.make(Maps.newHashMap(), map -> {
-        map.put("{}", Lists.newArrayList("DataVersion", "author", "size", "data", "entities", "palette", "palettes"));
-        map.put("{}.data.[].{}", Lists.newArrayList("pos", "state", "nbt"));
-        map.put("{}.entities.[].{}", Lists.newArrayList("blockPos", "pos"));
+    private static final Map<String, List<String>> KEY_ORDER = Util.make(new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(), map -> {
+        map.put("{}", one.tranic.t.utils.Collections.newArrayList("DataVersion", "author", "size", "data", "entities", "palette", "palettes"));
+        map.put("{}.data.[].{}", one.tranic.t.utils.Collections.newArrayList("pos", "state", "nbt"));
+        map.put("{}.entities.[].{}", one.tranic.t.utils.Collections.newArrayList("blockPos", "pos"));
     });
     private static final Set<String> NO_INDENTATION = Sets.newHashSet("{}.size.[]", "{}.data.[].{}", "{}.palette.[].{}", "{}.entities.[].{}");
     private static final Pattern SIMPLE_VALUE = Pattern.compile("[A-Za-z0-9._+-]+");
@@ -36,7 +36,7 @@ public class SnbtPrinterTagVisitor implements TagVisitor {
     private String result = "";
 
     public SnbtPrinterTagVisitor() {
-        this("    ", 0, Lists.newArrayList());
+        this("    ", 0, new it.unimi.dsi.fastutil.objects.ObjectArrayList<>());
     }
 
     public SnbtPrinterTagVisitor(String indentation, int depth, List<String> path) {
@@ -213,8 +213,8 @@ public class SnbtPrinterTagVisitor implements TagVisitor {
     }
 
     protected List<String> getKeys(CompoundTag tag) {
-        Set<String> set = Sets.newHashSet(tag.getAllKeys());
-        List<String> list = Lists.newArrayList();
+        Set<String> set = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(tag.getAllKeys());
+        List<String> list = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         List<String> list1 = KEY_ORDER.get(this.pathString());
         if (list1 != null) {
             for (String string : list1) {
diff --git a/net/minecraft/nbt/TagParser.java b/net/minecraft/nbt/TagParser.java
index c02a8042cfbe9f84cb7e68715188ae47feb92eb5..1af9f5653c275ddf85d15695249274c7deba27c4 100644
--- a/net/minecraft/nbt/TagParser.java
+++ b/net/minecraft/nbt/TagParser.java
@@ -246,7 +246,7 @@ public class TagParser {
     }
 
     private <T extends Number> List<T> readArray(TagType<?> arrayType, TagType<?> elementType) throws CommandSyntaxException {
-        List<T> list = Lists.newArrayList();
+        List<T> list = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
 
         while (this.reader.peek() != ']') {
             int cursor = this.reader.getCursor();
diff --git a/net/minecraft/nbt/TextComponentTagVisitor.java b/net/minecraft/nbt/TextComponentTagVisitor.java
index 5ae49516a3027c133e753810be62e44cbeff09fb..2147960310b62f04dccd4af24d15a44ceb5dd8c0 100644
--- a/net/minecraft/nbt/TextComponentTagVisitor.java
+++ b/net/minecraft/nbt/TextComponentTagVisitor.java
@@ -213,7 +213,7 @@ public class TextComponentTagVisitor implements TagVisitor {
             this.result.append("{");
             Collection<String> allKeys = tag.getAllKeys();
             if (LOGGER.isDebugEnabled()) {
-                List<String> list = Lists.newArrayList(tag.getAllKeys());
+                List<String> list = one.tranic.t.utils.Collections.newArrayList(tag.getAllKeys());
                 Collections.sort(list);
                 allKeys = list;
             }
