From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 21 Feb 2025 23:59:47 +0800
Subject: [PATCH] Modify EntityStorage


diff --git a/net/minecraft/world/level/chunk/storage/EntityStorage.java b/net/minecraft/world/level/chunk/storage/EntityStorage.java
index 2856206eafddfcbcc1b65408deda40357f43a6f8..c8de348f81d8c82ee8f423ff956d192608794362 100644
--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -93,7 +93,7 @@ public class EntityStorage implements EntityPersistentStorage<Entity> {
             }
         } else {
             ListTag listTag = new ListTag();
-            final java.util.Map<net.minecraft.world.entity.EntityType<?>, Integer> savedEntityCounts = new java.util.HashMap<>(); // Paper - Entity load/save limit per chunk
+            final java.util.Map<net.minecraft.world.entity.EntityType<?>, Integer> savedEntityCounts = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // Paper - Entity load/save limit per chunk
             entities.getEntities().forEach(entity -> {
                 // Paper start - Entity load/save limit per chunk
                 final EntityType<?> entityType = entity.getType();
