From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 25 Feb 2025 18:18:41 +0800
Subject: [PATCH] Modify network


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 2021142e677e028654f26cb9ecfc29e2aefc9825..7592dc8ac2b4f3df7da1a8619803cc7bde7b4d95 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -133,7 +133,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     protected final @Nullable io.papermc.paper.util.IntervalledCounter allPacketCounts = io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.allPackets.isEnabled() ? new io.papermc.paper.util.IntervalledCounter(
         (long)(io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.allPackets.interval() * 1.0e9)
     ) : null;
-    protected final java.util.Map<Class<? extends net.minecraft.network.protocol.Packet<?>>, io.papermc.paper.util.IntervalledCounter> packetSpecificLimits = new java.util.HashMap<>();
+    protected final java.util.Map<Class<? extends net.minecraft.network.protocol.Packet<?>>, io.papermc.paper.util.IntervalledCounter> packetSpecificLimits = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
 
     private boolean stopReadingPackets;
     private void killForPacketSpam() {
@@ -481,7 +481,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             if (!hasExtraPackets) {
                 this.pendingActions.add(new PacketSendAction(packet, listener, flush));
             } else {
-                final java.util.List<PacketSendAction> actions = new java.util.ArrayList<>(1 + extraPackets.size());
+                final java.util.List<PacketSendAction> actions = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(1 + extraPackets.size());
                 actions.add(new PacketSendAction(packet, null, false)); // Delay the future listener until the end of the extra packets
 
                 for (int i = 0, len = extraPackets.size(); i < len;) {
@@ -987,7 +987,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public void clearPacketQueue() {
         final net.minecraft.server.level.ServerPlayer player = getPlayer();
         // Folia start - region threading - connection fixes
-        java.util.List<Connection.PacketSendAction> queuedPackets = new java.util.ArrayList<>();
+        java.util.List<Connection.PacketSendAction> queuedPackets = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         // synchronise so that flushQueue does not poll values while the queue is being cleared
         synchronized (this.pendingActions) {
             Connection.WrappedConsumer consumer;
@@ -1016,7 +1016,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                 return null;
             }
 
-            final java.util.List<Packet<?>> ret = new java.util.ArrayList<>(1 + extra.size());
+            final java.util.List<Packet<?>> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(1 + extra.size());
             buildExtraPackets0(extra, ret);
             return ret;
         }
diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 9e321ef1c3d5803519b243685f4ee598dc0cf640..a42e79a4f88e34f263577e3a17949454f5db5321 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -28,7 +28,7 @@ public class ClientboundLevelChunkPacketData {
     private final byte[] buffer;
     private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
     // Paper start - Handle oversized block entities in chunks
-    private final java.util.List<net.minecraft.network.protocol.Packet<?>> extraPackets = new java.util.ArrayList<>();
+    private final java.util.List<net.minecraft.network.protocol.Packet<?>> extraPackets = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
     private static final int BLOCK_ENTITY_LIMIT = Integer.getInteger("Paper.excessiveTELimit", 750);
 
     public List<net.minecraft.network.protocol.Packet<?>> getExtraPackets() {
@@ -57,7 +57,7 @@ public class ClientboundLevelChunkPacketData {
             chunkPacketInfo.setBuffer(this.buffer);
         }
         extractChunkData(new FriendlyByteBuf(this.getWriteBuffer()), levelChunk, chunkPacketInfo);
-        this.blockEntitiesData = Lists.newArrayList();
+        this.blockEntitiesData = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         int totalTileEntities = 0; // Paper - Handle oversized block entities in chunks
 
         for (Entry<BlockPos, BlockEntity> entryx : levelChunk.getBlockEntities().entrySet()) {
