From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 15 Mar 2025 10:40:45 +0800
Subject: [PATCH] Modify some io.papermc


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index c260741a87513b89a5cc62c543fb9f990f86491e..0144887f1c04468442d6c3543087b8822dbd482b 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -154,7 +154,7 @@ public final class ActivationRange {
             bbByType[ActivationType.FLYING_MONSTER.ordinal()] = player.getBoundingBox().inflate(flyingActivationRange, worldHeight, flyingActivationRange); // Folia - threaded regions
             bbByType[ActivationType.VILLAGER.ordinal()] = player.getBoundingBox().inflate(villagerActivationRange, worldHeight, villagerActivationRange); // Folia - threaded regions
 
-            final java.util.List<Entity> entities = new java.util.ArrayList<>();  // Folia - region ticking - bypass getEntities thread check, we perform a check on the entities later
+            final java.util.List<Entity> entities = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();  // Folia - region ticking - bypass getEntities thread check, we perform a check on the entities later
             ((net.minecraft.server.level.ServerLevel)world).moonrise$getEntityLookup().getEntities((Entity)null, maxBB, entities, null); // Folia - region ticking - bypass getEntities thread check, we perform a check on the entities later
             final boolean tickMarkers = world.paperConfig().entities.markers.tick;
             for (final Entity entity : entities) {
diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index 5a76c93eada8db35b1ddbb562ccfbd2f0d35f0ca..73241fee54ba84c4690e33b4dc7ae8a7f1420379 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -108,9 +108,9 @@ public final class RedstoneWireTurbo {
      * information flow.  This makes the implementation non-orientational.  Only when
      * this direction is ambiguous is randomness applied (intentionally).
      */
-    private List<UpdateNode> updateQueue0 = Lists.newArrayList();
-    private List<UpdateNode> updateQueue1 = Lists.newArrayList();
-    private List<UpdateNode> updateQueue2 = Lists.newArrayList();
+    private List<UpdateNode> updateQueue0 = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+    private List<UpdateNode> updateQueue1 = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+    private List<UpdateNode> updateQueue2 = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
 
     public RedstoneWireTurbo(RedStoneWireBlock wire) {
         this.wire = wire;
diff --git a/io/papermc/paper/util/PoiAccess.java b/io/papermc/paper/util/PoiAccess.java
index f39294b1f83c4022be5ced4da781103a1eee2daf..278edadac9860f5bb5c6432a45bb8cab7f147155 100644
--- a/io/papermc/paper/util/PoiAccess.java
+++ b/io/papermc/paper/util/PoiAccess.java
@@ -126,7 +126,7 @@ public final class PoiAccess {
                                                    final PoiManager.Occupancy occupancy,
                                                    final boolean load,
                                                    final Set<BlockPos> ret) {
-        final Set<BlockPos> positions = new HashSet<>();
+        final Set<BlockPos> positions = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
         // pos predicate is last thing that runs before adding to ret.
         final Predicate<BlockPos> newPredicate = (final BlockPos pos) -> {
             if (positionPredicate != null && !positionPredicate.test(pos)) {
@@ -135,7 +135,7 @@ public final class PoiAccess {
             return positions.add(pos.immutable());
         };
 
-        final List<PoiRecord> toConvert = new ArrayList<>();
+        final List<PoiRecord> toConvert = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findClosestPoiDataRecords(
                 poiStorage, villagePlaceType, newPredicate, sourcePosition, range, maxDistanceSquared, occupancy, load, toConvert
         );
@@ -156,7 +156,7 @@ public final class PoiAccess {
                                                      final double maxDistanceSquared,
                                                      final PoiManager.Occupancy occupancy,
                                                      final boolean load) {
-        final List<PoiRecord> ret = new ArrayList<>();
+        final List<PoiRecord> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findClosestPoiDataRecords(
             poiStorage, villagePlaceType, positionPredicate, sourcePosition, range, maxDistanceSquared, occupancy, load, ret
         );
@@ -174,7 +174,7 @@ public final class PoiAccess {
                                                      final double maxDistanceSquared,
                                                      final PoiManager.Occupancy occupancy,
                                                      final boolean load) {
-        final List<PoiRecord> ret = new ArrayList<>();
+        final List<PoiRecord> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findClosestPoiDataRecords(
                 poiStorage, villagePlaceType, predicate, sourcePosition, range, maxDistanceSquared, occupancy, load, ret
         );
@@ -209,7 +209,7 @@ public final class PoiAccess {
                                                  final List<PoiRecord> ret) {
         final Predicate<? super PoiRecord> occupancyFilter = occupancy.getTest();
 
-        final List<PoiRecord> closestRecords = new ArrayList<>();
+        final List<PoiRecord> closestRecords = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         double closestDistanceSquared = maxDistanceSquared;
 
         final int lowerX = Mth.floor(sourcePosition.getX() - range) >> 4;
@@ -400,7 +400,7 @@ public final class PoiAccess {
                                                final boolean load,
                                                final int max,
                                                final List<Pair<Holder<PoiType>, BlockPos>> ret) {
-        final Set<BlockPos> positions = new HashSet<>();
+        final Set<BlockPos> positions = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
         // pos predicate is last thing that runs before adding to ret.
         final Predicate<BlockPos> newPredicate = (final BlockPos pos) -> {
             if (positionPredicate != null && !positionPredicate.test(pos)) {
@@ -409,7 +409,7 @@ public final class PoiAccess {
             return positions.add(pos.immutable());
         };
 
-        final List<PoiRecord> toConvert = new ArrayList<>();
+        final List<PoiRecord> toConvert = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findNearestPoiRecords(
                 poiStorage, villagePlaceType, newPredicate, sourcePosition, range, maxDistanceSquared, occupancy, load, max, toConvert
         );
@@ -429,7 +429,7 @@ public final class PoiAccess {
                                                  final double maxDistanceSquared,
                                                  final PoiManager.Occupancy occupancy,
                                                  final boolean load) {
-        final List<PoiRecord> ret = new ArrayList<>();
+        final List<PoiRecord> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findNearestPoiRecords(
                 poiStorage, villagePlaceType, positionPredicate, sourcePosition, range, maxDistanceSquared, occupancy, load,
                 1, ret
@@ -578,7 +578,7 @@ public final class PoiAccess {
                     }
 
                     closestRecords.computeIfAbsent(dataRange, (final double unused) -> {
-                        return new ArrayList<>();
+                        return new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
                     }).add(poiData);
 
                     if (++totalRecords >= max) {
@@ -607,7 +607,7 @@ public final class PoiAccess {
             }
         }
 
-        final List<PoiRecord> closestRecordsUnsorted = new ArrayList<>();
+        final List<PoiRecord> closestRecordsUnsorted = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
 
         // we're done here, so now just flatten the map and sort it.
 
@@ -683,7 +683,7 @@ public final class PoiAccess {
                                            final boolean load,
                                            final int max,
                                            final List<Pair<Holder<PoiType>, BlockPos>> ret) {
-        final Set<BlockPos> positions = new HashSet<>();
+        final Set<BlockPos> positions = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
         // pos predicate is last thing that runs before adding to ret.
         final Predicate<BlockPos> newPredicate = (final BlockPos pos) -> {
             if (positionPredicate != null && !positionPredicate.test(pos)) {
@@ -692,7 +692,7 @@ public final class PoiAccess {
             return positions.add(pos.immutable());
         };
 
-        final List<PoiRecord> toConvert = new ArrayList<>();
+        final List<PoiRecord> toConvert = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findAnyPoiRecords(
                 poiStorage, villagePlaceType, newPredicate, sourcePosition, range, occupancy, load, max, toConvert
         );
@@ -709,7 +709,7 @@ public final class PoiAccess {
                                              final int range, // distance on x y z axis
                                              final PoiManager.Occupancy occupancy,
                                              final boolean load) {
-        final List<PoiRecord> ret = new ArrayList<>();
+        final List<PoiRecord> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
         findAnyPoiRecords(poiStorage, villagePlaceType, positionPredicate, sourcePosition, range, occupancy, load, 1, ret);
         return ret.isEmpty() ? null : ret.get(0);
     }
