From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Thu, 13 Mar 2025 18:09:45 +0800
Subject: [PATCH] Modify yggdrasil


diff --git a/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java b/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
index 55dac5edf694b3bf82b475a71e3524a1bce98882..d7681deca079e2785246665948fa46dfb16b42f4 100644
--- a/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
+++ b/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
@@ -39,15 +39,17 @@ public class YggdrasilGameProfileRepository implements GameProfileRepository {
 
     @Override
     public void findProfilesByNames(final String[] names, final ProfileLookupCallback callback) {
-        final Set<String> criteria = Arrays.stream(names)
-            .filter(name -> !Strings.isNullOrEmpty(name))
-            .collect(Collectors.toSet());
+        final Set<String> criteria = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
+        for (String name : names) if (!Strings.isNullOrEmpty(name)) criteria.add(name);
 
         final int page = 0;
         boolean hasRequested = false; // Paper - Don't sleep after profile lookups if not needed
 
         for (final List<String> request : Iterables.partition(criteria, ENTRIES_PER_PAGE)) {
-            final List<String> normalizedRequest = request.stream().map(YggdrasilGameProfileRepository::normalizeName).toList();
+            final List<String> normalizedRequest = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+            for (String item : request) {
+                normalizedRequest.add(YggdrasilGameProfileRepository.normalizeName(item));
+            }
 
             int failCount = 0;
             boolean failed;
diff --git a/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java b/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java
index 3eaa893b44e4370bcb04582d08958645a6c57316..62f3b7e4d22a7c7e2d4087644d623f7f0c13be5d 100644
--- a/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java
+++ b/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java
@@ -74,9 +74,9 @@ public class YggdrasilServicesKeyInfo implements ServicesKeyInfo {
         if (keys == null) {
             return List.of();
         }
-        return keys.stream()
-            .map(data -> parse(data.publicKey.array()))
-            .toList();
+        List<ServicesKeyInfo> keyInfoList = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        for (KeyData data : keys) keyInfoList.add(parse(data.publicKey.array()));
+        return keyInfoList;
     }
 
     public static ServicesKeySet get(final URL url, final MinecraftClient client) {
