From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 10 Feb 2025 21:43:41 +0800
Subject: [PATCH] Pufferfish: Reduce projectile chunk loading


diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index 6e1de9252ac1236e0c5196e2511c72c0d7eb3502..8b2959ed08364cc96341d36e53161521ee143a2b 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -49,6 +49,41 @@ public abstract class Projectile extends Entity implements TraceableEntity {
         super(entityType, level);
     }
 
+    // Pufferfish start
+    private static final java.lang.ThreadLocal<Long> loadedThisTick = java.lang.ThreadLocal.withInitial(() -> 0L);
+    private static final java.lang.ThreadLocal<Long> loadedTick = java.lang.ThreadLocal.withInitial(() -> 0L);
+
+    private int loadedLifetime = 0;
+    @Override
+    public void setPos(double x, double y, double z) {
+        if (one.tranic.sewlia.config.mod.optimization.projectile_loads.Enable.value && io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegion() != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) {
+            long currentTick = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegion().getData().getCurrentTick();
+
+            if (loadedTick.get() != currentTick) {
+                loadedTick.set(currentTick);
+                loadedThisTick.set(0L);
+            }
+
+            int previousX = Mth.floor(this.getX()) >> 4, previousZ = Mth.floor(this.getZ()) >> 4;
+            int newX = Mth.floor(x) >> 4, newZ = Mth.floor(z) >> 4;
+            if (previousX != newX || previousZ != newZ) {
+                boolean isLoaded = ((net.minecraft.server.level.ServerChunkCache) this.level().getChunkSource()).getChunkAtIfLoadedImmediately(newX, newZ) != null;
+                if (!isLoaded) {
+                    if (Projectile.loadedThisTick.get() > one.tranic.sewlia.config.mod.optimization.projectile_loads.MaxTick.value) {
+                        if (++this.loadedLifetime > one.tranic.sewlia.config.mod.optimization.projectile_loads.Max.value) {
+                            this.discard();
+                        }
+                        return;
+                    }
+                    Projectile.loadedThisTick.set(Projectile.loadedThisTick.get() + 1);
+                }
+            }
+            return;
+        }
+        super.setPosRaw(x, y, z);
+    }
+    // Pufferfish end
+
     public void setOwner(@Nullable Entity owner) {
         if (owner != null) {
             this.ownerUUID = owner.getUUID();
