From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Wed, 23 Oct 2024 23:10:48 +0800
Subject: [PATCH] Leaves: Revert raid changes


diff --git a/net/minecraft/world/effect/BadOmenMobEffect.java b/net/minecraft/world/effect/BadOmenMobEffect.java
index 80f17f33f670018240c854df589cf90cdeab6e70..0fb30899c873bcaf83c18447969c3ba85694a10b 100644
--- a/net/minecraft/world/effect/BadOmenMobEffect.java
+++ b/net/minecraft/world/effect/BadOmenMobEffect.java
@@ -22,6 +22,11 @@ class BadOmenMobEffect extends MobEffect {
             && !serverPlayer.isSpectator()
             && level.getDifficulty() != Difficulty.PEACEFUL
             && level.isVillage(serverPlayer.blockPosition())) {
+            // Leaves start - Revert raid changes
+            if (one.tranic.sewlia.config.mod.feature.raid.AllowBadOmenTriggerRaid.value) {
+                return level.getRaids().createOrExtendRaid(serverPlayer, serverPlayer.blockPosition()) != null;
+            }
+            // Leaves end - Revert raid changes
             Raid raidAt = level.getRaidAt(serverPlayer.blockPosition());
             if (raidAt == null || raidAt.getRaidOmenLevel() < raidAt.getMaxRaidOmenLevel()) {
                 serverPlayer.addEffect(new MobEffectInstance(MobEffects.RAID_OMEN, 600, amplifier));
diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index 7c385baae81b9a987c0e1e4deb017884600331bc..ce45ac6733861da2f202066041180bbb50811767 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -16,8 +16,11 @@ import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
+import net.minecraft.util.Mth;
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.effect.MobEffectInstance;
+import net.minecraft.world.effect.MobEffects;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySpawnReason;
 import net.minecraft.world.entity.EntityType;
@@ -31,10 +34,13 @@ import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.entity.ai.village.poi.PoiManager;
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
+import net.minecraft.world.entity.animal.Wolf;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.monster.AbstractIllager;
 import net.minecraft.world.entity.monster.PatrollingMonster;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.pathfinder.Path;
@@ -125,6 +131,38 @@ public abstract class Raider extends PatrollingMonster {
 
                 currentRaid.removeFromRaid(this, false);
             }
+
+            // Leaves start - Revert raid changes
+            if (this.level() instanceof ServerLevel serverLevel) {
+                if (one.tranic.sewlia.config.mod.feature.raid.GiveBadOmenWhenKillPatrolLeader.value && this.isPatrolLeader() && raid == null && serverLevel.getRaidAt(this.blockPosition()) == null) {
+                    ItemStack itemstack = this.getItemBySlot(EquipmentSlot.HEAD);
+                    Player entityhuman = null;
+
+                    if (entity instanceof Player) {
+                        entityhuman = (Player) entity;
+                    } else if (entity instanceof Wolf entitywolf) {
+                        LivingEntity entityliving = entitywolf.getOwner();
+                        if (entitywolf.isTame() && entityliving instanceof Player)
+                            entityhuman = (Player) entityliving;
+                    }
+
+                    if (!itemstack.isEmpty() && ItemStack.matches(itemstack, Raid.getOminousBannerInstance(this.registryAccess().lookupOrThrow(Registries.BANNER_PATTERN))) && entityhuman != null) {
+                        MobEffectInstance mobeffect = entityhuman.getEffect(MobEffects.BAD_OMEN);
+                        int i = 1;
+
+                        if (mobeffect != null) {
+                            i += mobeffect.getAmplifier();
+                            entityhuman.removeEffectNoUpdate(MobEffects.BAD_OMEN);
+                        } else --i;
+
+                        i = Mth.clamp(i, 0, 4);
+                        MobEffectInstance mobeffect1 = new MobEffectInstance(MobEffects.BAD_OMEN, 120000, i, false, false, true);
+                        if (!serverLevel.getGameRules().getBoolean(GameRules.RULE_DISABLE_RAIDS))
+                            entityhuman.addEffect(mobeffect1, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
+                    }
+                }
+            }
+            // Leaves end - Revert raid changes
         }
 
         super.die(cause);
