From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 12:00:59 +0800
Subject: [PATCH] Leaves: Fix villagers dont release memory


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index e1c61c9a033707d6adbc713f43edccb2aa849656..06f4176c3ccd2b95e934071192fced699c499364 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4640,7 +4640,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this;
     }
 
-    private Entity teleportCrossDimension(ServerLevel level, TeleportTransition teleportTransition) {
+    protected Entity teleportCrossDimension(ServerLevel level, TeleportTransition teleportTransition) { // Leaves - private -> protected
         List<Entity> passengers = this.getPassengers();
         List<Entity> list = new ArrayList<>(passengers.size());
         this.ejectPassengers();
diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index db235101be529b50cb6925d2b7ff015bd2a7cf84..19ee59b124cd2f866fd0b5e73b42c793d461e98d 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -1030,4 +1030,19 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         Optional<Long> memory = this.brain.getMemory(MemoryModuleType.LAST_SLEPT);
         return memory.filter(_long -> gameTime - _long < 24000L).isPresent();
     }
+
+    // Leaves start - fixes a memory leak when villagers get moved to another world
+    @Override
+    public Entity teleportCrossDimension(ServerLevel world, net.minecraft.world.level.portal.TeleportTransition transition) {
+        if (one.tranic.sewlia.config.mod.fix.VillagersDontReleaseMemory.value) {
+            this.releaseAllPois();
+            this.getBrain().eraseMemory(MemoryModuleType.HOME);
+            this.getBrain().eraseMemory(MemoryModuleType.JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.POTENTIAL_JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.MEETING_POINT);
+            this.refreshBrain(transition.newLevel());
+        }
+        return super.teleportCrossDimension(world, transition);
+    }
+    // Leaves end - fixes a memory leak when villagers get moved to another world
 }
