From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 26 Oct 2024 19:52:46 +0800
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index 571534b42cd9c33d6a7bb6fe3bf3a28e33f8e5de..8940c72e293776a0f1279b4fc38c27d22f012d57 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -12,8 +12,8 @@ java {
 val annotationsVersion = "24.1.0"
 val bungeeCordChatVersion = "1.20-R0.2"
 val adventureVersion = "4.17.0"
-val slf4jVersion = "2.0.9"
-val log4jVersion = "2.17.1"
+val slf4jVersion = "2.0.16"
+val log4jVersion = "2.23.1"
 val apiAndDocs: Configuration by configurations.creating {
     attributes {
         attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.DOCUMENTATION))
@@ -48,7 +48,7 @@ dependencies {
         exclude("com.google.guava", "guava")
     }
     // Paper - adventure
-    api("org.yaml:snakeyaml:2.2")
+    api("org.yaml:snakeyaml:2.3")
     api("org.joml:joml:1.10.8") {
         isTransitive = false // https://github.com/JOML-CI/JOML/issues/352
     }
@@ -86,6 +86,10 @@ dependencies {
     // Paper end
     api("org.jspecify:jspecify:1.0.0") // Paper - add jspecify
 
+    // Sewlia start
+    api("net.intelie.tinymap:tinymap:0.7")
+    // Sewlia end
+
     testImplementation("org.apache.commons:commons-lang3:3.12.0")
     testImplementation("org.junit.jupiter:junit-jupiter:5.10.2")
     testImplementation("org.hamcrest:hamcrest:2.2")
@@ -168,7 +172,7 @@ tasks.withType<Javadoc> {
     options.isDocFilesSubDirs = true
     options.links(
         "https://guava.dev/releases/33.3.1-jre/api/docs/",
-        "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/",
+        "https://javadoc.io/doc/org.yaml/snakeyaml/2.3/",
         "https://javadoc.io/doc/org.jetbrains/annotations/$annotationsVersion/", // Paper - we don't want Java 5 annotations
         // "https://javadoc.io/doc/net.md-5/bungeecord-chat/$bungeeCordChatVersion/", // Paper - don't link to bungee chat
         // Paper start - add missing javadoc links
