From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 3 Nov 2024 10:39:20 +0800
Subject: [PATCH] Ignore compile time warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index 4dd8c0d2de9e9ad14ce16221c3d351af444fd0c2..457f7f629971a2d0a3b3c3b34f77656c97103c3f 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -204,9 +204,16 @@ tasks.test {
 // Paper start - compile tests with -parameters for better junit parameterized test names
 tasks.compileTestJava {
     options.compilerArgs.add("-parameters")
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
 }
 // Paper end
 
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
+}
+
 // Paper start
 val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.ScanJarForBadCalls::class) {
     badAnnotations.add("Lio/papermc/paper/annotation/DoNotUse;")
