From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 3 Nov 2024 10:39:20 +0800
Subject: [PATCH] Ignore compile time warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index bad513122bcffe01d0d46898f7366fffeb8e4f26..2ed9d0dbf2d69084524332fa88bb8c619cfffd60 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -201,9 +201,16 @@ tasks.test {
 // Paper start - compile tests with -parameters for better junit parameterized test names
 tasks.compileTestJava {
     options.compilerArgs.add("-parameters")
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
 }
 // Paper end
 
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
+}
+
 // Paper start
 val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.ScanJarForBadCalls::class) {
     badAnnotations.add("Lio/papermc/paper/annotation/DoNotUse;")
