From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 3 Nov 2024 10:39:20 +0800
Subject: [PATCH] Ignore compile time warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index 428c62a884999345126eca20ce8c5a5b27561199..c1f2a8ab25de9bbb85fb97218d178d65d8d6da22 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -225,9 +225,16 @@ tasks.test {
 // Paper start - compile tests with -parameters for better junit parameterized test names
 tasks.compileTestJava {
     options.compilerArgs.add("-parameters")
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
 }
 // Paper end
 
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
+}
+
 // Paper start
 val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.ScanJarForBadCalls::class) {
     badAnnotations.add("Lio/papermc/paper/annotation/DoNotUse;")
