From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 3 Nov 2024 10:39:20 +0800
Subject: [PATCH] Ignore compile time warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index 8940c72e293776a0f1279b4fc38c27d22f012d57..435322f62dae8b1a5cb40d43e8c8c3bc779af723 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -227,9 +227,16 @@ tasks.test {
 // Paper start - compile tests with -parameters for better junit parameterized test names
 tasks.compileTestJava {
     options.compilerArgs.add("-parameters")
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
 }
 // Paper end
 
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation") // Sewlia
+    options.isWarnings = false // Sewlia
+}
+
 // Paper start
 val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.ScanJarForBadCalls::class) {
     badAnnotations.add("Lio/papermc/paper/annotation/DoNotUse;")
