From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 30 Oct 2024 17:23:42 +0800
Subject: [PATCH] Sewlia Config Key


diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/TpsCatchup.java b/src/main/java/one/tranic/sewlia/config/mod/feature/TpsCatchup.java
new file mode 100644
index 0000000000000000000000000000000000000000..aa1258afdcddce20875aa2c8fa687e46a8dc2df3
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/TpsCatchup.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class TpsCatchup {
+    @Comments({"If this setting is true, the server will run faster after a lag spike in",
+            "an attempt to maintain 20 TPS. This option (defaults to true per",
+            "spigot/paper) can cause mobs to move fast after a lag spike."})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Designation.java b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Designation.java
new file mode 100644
index 0000000000000000000000000000000000000000..f9f11e48f4d214636652f7e9907947f1ab259777
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Designation.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Designation {
+    @DisableReload
+    @Comment("Manually specify a repository")
+    public static String value = "";
+}
\ No newline at end of file
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..25b2632665eb933feff710bb30a015341bb381cb
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Enable.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Enable {
+    @DisableReload
+    @Comment("Maven Central Repository Acceleration")
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..aa673e7c22f4780bec656d1506b48d7b0eee67be
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class UserRepository {
+    @DisableReload
+    @Comment("Users add a central repository mirror")
+    public static String value = "";
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/POIAccessOffRegion.java b/src/main/java/one/tranic/sewlia/config/mod/fix/POIAccessOffRegion.java
new file mode 100644
index 0000000000000000000000000000000000000000..bda476a96f551ea1d303d85d425968e112f118ee
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/POIAccessOffRegion.java
@@ -0,0 +1,11 @@
+package one.tranic.sewlia.config.mod.fix;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class POIAccessOffRegion {
+    @Comments({"The POIManager of folia has something which has not been patched",
+            "for regionized ticking and these would trigger the async catcher",
+            "and make the server crash.If you would like to prevent it and didn't",
+            "mind the side effect(currently unknown), you can enable this"})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/SpectorTeleportation.java b/src/main/java/one/tranic/sewlia/config/mod/fix/SpectorTeleportation.java
new file mode 100644
index 0000000000000000000000000000000000000000..2600c7215a47df59d5cc1220fbf59bdf34c715ca
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/SpectorTeleportation.java
@@ -0,0 +1,11 @@
+package one.tranic.sewlia.config.mod.fix;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class SpectorTeleportation {
+    @Comments({"The teleportation of spector players would call absMoveTo directly.",
+            "And when the camera teleported to another region,this would call absMoveTo",
+            "to let the spector player move to another region without any checks, which ",
+            "would trigger the async catcher and crash the server"})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/TaskQueue.java b/src/main/java/one/tranic/sewlia/config/mod/fix/TaskQueue.java
new file mode 100644
index 0000000000000000000000000000000000000000..51c24fa14f56f5f19fa528189a6a01e40b00879f
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/TaskQueue.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.fix;
+
+public class TaskQueue {
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/VillagersDontReleaseMemory.java b/src/main/java/one/tranic/sewlia/config/mod/fix/VillagersDontReleaseMemory.java
new file mode 100644
index 0000000000000000000000000000000000000000..80c80347b9c4a8a00b6c52b92ac0ef5f0e79c4b3
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/VillagersDontReleaseMemory.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.fix;
+
+import one.tranic.sewlia.config.annotation.Comment;
+
+public class VillagersDontReleaseMemory {
+    @Comment("test comment")
+    public static Boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Debug.java b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Debug.java
new file mode 100644
index 0000000000000000000000000000000000000000..90260a1b65a0ca88d84b82ab6a9e92bfe6ae72a8
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Debug.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.fix.anvil_drop;
+
+import one.tranic.sewlia.config.annotation.Comment;
+
+public class Debug {
+    @Comment("Whether to print detailed information to the console when this issue is triggered")
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..8234a65d2aa6b5b3d607cec95453efe41e9f82d9
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Enable.java
@@ -0,0 +1,9 @@
+package one.tranic.sewlia.config.mod.fix.anvil_drop;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class Enable {
+    @Comments({"This fix will attempt to prevent the bug in https://github.com/PaperMC/Folia/issues/217, ",
+            "it may cause some unexpected effects but should be harmless."})
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..8fac641a45325f0f5b7c03497e1ce6abed731e94
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/Enable.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.fix.entity_moving;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class Enable {
+    @Comments({"A simple fix of a issue on folia.",
+            "Some times the entity would have a large moment that cross the different tick regions and it would",
+            "make the server crashed)  but sometimes it might doesn't work."})
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/WarnOnDetected.java b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/WarnOnDetected.java
new file mode 100644
index 0000000000000000000000000000000000000000..54d5ac1138baa45f35e6b110112dcae022f9e873
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/WarnOnDetected.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.fix.entity_moving;
+
+public class WarnOnDetected {
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/RunExtraTasksDuringWaiting.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/RunExtraTasksDuringWaiting.java
new file mode 100644
index 0000000000000000000000000000000000000000..7549383deabee9b461becdd308d9c8708bf4c91d
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/RunExtraTasksDuringWaiting.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.optimization;
+
+public class RunExtraTasksDuringWaiting {
+    public static boolean value = false;
+}
