From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 30 Oct 2024 17:23:42 +0800
Subject: [PATCH] Sewlia Config Key


diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/TpsCatchup.java b/src/main/java/one/tranic/sewlia/config/mod/feature/TpsCatchup.java
new file mode 100644
index 0000000000000000000000000000000000000000..47c59c00e5232ab551dc485ab34280e297f298b8
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/TpsCatchup.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature;
+
+import one.tranic.sewlia.config.annotation.Comments;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class TpsCatchup {
+    @Comments({"If this setting is true, the server will run faster after a lag spike in",
+            "an attempt to maintain 20 TPS. This option (defaults to true per",
+            "spigot/paper) can cause mobs to move fast after a lag spike."})
+    @DisableReload
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Designation.java b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Designation.java
new file mode 100644
index 0000000000000000000000000000000000000000..f9f11e48f4d214636652f7e9907947f1ab259777
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Designation.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Designation {
+    @DisableReload
+    @Comment("Manually specify a repository")
+    public static String value = "";
+}
\ No newline at end of file
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..25b2632665eb933feff710bb30a015341bb381cb
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/Enable.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Enable {
+    @DisableReload
+    @Comment("Maven Central Repository Acceleration")
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..aa673e7c22f4780bec656d1506b48d7b0eee67be
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/maven/UserRepository.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.maven;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class UserRepository {
+    @DisableReload
+    @Comment("Users add a central repository mirror")
+    public static String value = "";
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/AsyncScheduler.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/AsyncScheduler.java
new file mode 100644
index 0000000000000000000000000000000000000000..fa69e4abc7780f721b2b156d8e3165cfdca03620
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/AsyncScheduler.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class AsyncScheduler {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/CraftScheduler.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/CraftScheduler.java
new file mode 100644
index 0000000000000000000000000000000000000000..b1d544475ea86d7035bbffb3dd3d92f6d850fb1f
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/CraftScheduler.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class CraftScheduler {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/ServerMain.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/ServerMain.java
new file mode 100644
index 0000000000000000000000000000000000000000..7e23ddb5b13355aa5728ce80dfd001247e09db8d
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/bukkit/ServerMain.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ServerMain {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable, final String name) {
+        return value ? Thread.ofVirtual().name(name).unstarted(runnable) : new Thread(runnable, name);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/Client.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/Client.java
new file mode 100644
index 0000000000000000000000000000000000000000..cde297f3c3ce4bf7b98ec4a9310fdb98ca83c0ea
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/Client.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.chase;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Client {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable, final String name) {
+        return value ? Thread.ofVirtual().name(name).unstarted(runnable) : new Thread(runnable, name);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/ServerAcceptor.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/ServerAcceptor.java
new file mode 100644
index 0000000000000000000000000000000000000000..1b82eaf05d07c25189f2b7cf88d1b35ec841467c
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/ServerAcceptor.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.chase;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ServerAcceptor {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable, final String name) {
+        return value ? Thread.ofVirtual().name(name).unstarted(runnable) : new Thread(runnable, name);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/ServerSender.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/ServerSender.java
new file mode 100644
index 0000000000000000000000000000000000000000..1e5ebedb1e5d4a01a3c82933c292064da740615e
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/chase/ServerSender.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.chase;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ServerSender {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable, final String name) {
+        return value ? Thread.ofVirtual().name(name).unstarted(runnable) : new Thread(runnable, name);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/folia/AsyncScheduler.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/folia/AsyncScheduler.java
new file mode 100644
index 0000000000000000000000000000000000000000..448760cfc9f70ea7e1bda12d6fa7841d9910f9bf
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/folia/AsyncScheduler.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.folia;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class AsyncScheduler {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable) {
+        return value ? Thread.ofVirtual().unstarted(runnable) : new Thread(runnable);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/AsyncChat.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/AsyncChat.java
new file mode 100644
index 0000000000000000000000000000000000000000..854534a080742b46a1293f98607f5d69b8e71467
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/AsyncChat.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class AsyncChat {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/AsyncDebug.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/AsyncDebug.java
new file mode 100644
index 0000000000000000000000000000000000000000..6005d96b0c3a7f26f55d95dc29c0655a06707973
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/AsyncDebug.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class AsyncDebug {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable r) {
+        return value ? Thread.ofVirtual().unstarted(r) : new Thread(r);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Console.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Console.java
new file mode 100644
index 0000000000000000000000000000000000000000..b8e0b5075e990ad74ec16165aebad0d8bef31860
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Console.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Console {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final String name, final Runnable r) {
+        return value ? Thread.ofVirtual().name(name).unstarted(r) : new Thread(r, name);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Datafixer.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Datafixer.java
new file mode 100644
index 0000000000000000000000000000000000000000..151af73dd502aa89b02bb828aedd438722c2369f
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Datafixer.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class Datafixer {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Gui.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Gui.java
new file mode 100644
index 0000000000000000000000000000000000000000..f4deb7fb7987f0d94599f1e790ed8fadaa621e12
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/Gui.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Gui {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable r) {
+        return value ? Thread.ofVirtual().unstarted(r) : new Thread(r);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/NamedThread.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/NamedThread.java
new file mode 100644
index 0000000000000000000000000000000000000000..b2d3d47ea7b458b418c600a532d4dd77306a65d2
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/NamedThread.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class NamedThread {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(@Nullable ThreadGroup group, Runnable task, @NotNull String name, long stackSize) {
+        return value ? Thread.ofVirtual().name(name).unstarted(task) : new Thread(group, task, name, stackSize);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/ServerWatchdog.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/ServerWatchdog.java
new file mode 100644
index 0000000000000000000000000000000000000000..4334821efb7a509c4672cf55e791525fccef3788
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/ServerWatchdog.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ServerWatchdog {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable r) {
+        return value ? Thread.ofVirtual().unstarted(r) : new Thread(r);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/DimensionData.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/DimensionData.java
new file mode 100644
index 0000000000000000000000000000000000000000..4571dc1fe87ff932f81636b34da308d5496b8b6b
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/DimensionData.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class DimensionData {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/Noise.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/Noise.java
new file mode 100644
index 0000000000000000000000000000000000000000..229cced655e7de62c68eb95c2d17ee0d0acec912
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/Noise.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Noise {
+    @DisableReload
+    @Comment("Scheduling noise generation to virtual threads may lead to some unpredictable consequences, so enable it with caution.")
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/Worker.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/Worker.java
new file mode 100644
index 0000000000000000000000000000000000000000..d22f244bf1ef047b18c6e3e6581277c4c19f00d4
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/Worker.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Worker {
+    @DisableReload
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/WorldUpgrader.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/WorldUpgrader.java
new file mode 100644
index 0000000000000000000000000000000000000000..1a66347a61c952a8392eab167e9ad5af39659e54
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/WorldUpgrader.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class WorldUpgrader {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/ChatFilterClient.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/ChatFilterClient.java
new file mode 100644
index 0000000000000000000000000000000000000000..3df1bc7e5703e1805b086cd39595ec46de64ccc4
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/ChatFilterClient.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ChatFilterClient {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable) {
+        return value ? Thread.ofVirtual().unstarted(runnable) : new Thread(runnable);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/Download.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/Download.java
new file mode 100644
index 0000000000000000000000000000000000000000..d8bdae9408fc366e8cedb9d94a445326d455b754
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/Download.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Download {
+    @DisableReload
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/ProfileLookup.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/ProfileLookup.java
new file mode 100644
index 0000000000000000000000000000000000000000..94a97342f21b2449617e43472836cfb20e78f76a
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/ProfileLookup.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ProfileLookup {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable runnable) {
+        return value ? Thread.ofVirtual().unstarted(runnable) : new Thread(runnable);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/RconGeneric.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/RconGeneric.java
new file mode 100644
index 0000000000000000000000000000000000000000..e6d4f97404d538780cecd609b5aab59707924f90
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/RconGeneric.java
@@ -0,0 +1,12 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class RconGeneric {
+    @DisableReload
+    public static boolean value = false;
+
+    public static Thread getThread(final Runnable r, final String name) {
+        return value ? Thread.ofVirtual().name(name).unstarted(r) : new Thread(r, name);
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/YggdrasilServices.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/YggdrasilServices.java
new file mode 100644
index 0000000000000000000000000000000000000000..77e180732b9cc7c2fbd444cf2ae6257dfb6cd247
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/network/YggdrasilServices.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class YggdrasilServices {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/Client.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/Client.java
new file mode 100644
index 0000000000000000000000000000000000000000..9c39bee4ef318f7f0a6de4d58ff70c3d3228e7b9
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/Client.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.netty;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class Client {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/EpollClient.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/EpollClient.java
new file mode 100644
index 0000000000000000000000000000000000000000..1e53f37d50f4a4c2f271c4bb43166ca15439dc49
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/EpollClient.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.netty;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class EpollClient {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/EpollServer.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/EpollServer.java
new file mode 100644
index 0000000000000000000000000000000000000000..ea5094b61e3208f083f9533ac7885453dd099808
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/EpollServer.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.netty;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class EpollServer {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/LocalClient.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/LocalClient.java
new file mode 100644
index 0000000000000000000000000000000000000000..f4b429c7b9575f2cf216e4bb292ef6c037f613f5
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/LocalClient.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.netty;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class LocalClient {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/Server.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/Server.java
new file mode 100644
index 0000000000000000000000000000000000000000..cb3d5cda54a4ed6085e95ed48f17823cabe0021d
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/netty/Server.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.netty;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class Server {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/AsyncCommandBuilder.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/AsyncCommandBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..42351217e6786cd92e91700478b3b6aa734d4f56
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/AsyncCommandBuilder.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.paper;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class AsyncCommandBuilder {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/AsyncTaskHandler.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/AsyncTaskHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..bc7a08d1eae8597d9c7357f51576d6e9213189b8
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/AsyncTaskHandler.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.paper;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class AsyncTaskHandler {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/PluginRemapper.java b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/PluginRemapper.java
new file mode 100644
index 0000000000000000000000000000000000000000..7a8aaa9962832a5b7a429dbcb84bb15fb3734422
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/PluginRemapper.java
@@ -0,0 +1,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.paper;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class PluginRemapper {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/POIAccessOffRegion.java b/src/main/java/one/tranic/sewlia/config/mod/fix/POIAccessOffRegion.java
new file mode 100644
index 0000000000000000000000000000000000000000..bda476a96f551ea1d303d85d425968e112f118ee
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/POIAccessOffRegion.java
@@ -0,0 +1,11 @@
+package one.tranic.sewlia.config.mod.fix;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class POIAccessOffRegion {
+    @Comments({"The POIManager of folia has something which has not been patched",
+            "for regionized ticking and these would trigger the async catcher",
+            "and make the server crash.If you would like to prevent it and didn't",
+            "mind the side effect(currently unknown), you can enable this"})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/SpectorTeleportation.java b/src/main/java/one/tranic/sewlia/config/mod/fix/SpectorTeleportation.java
new file mode 100644
index 0000000000000000000000000000000000000000..2600c7215a47df59d5cc1220fbf59bdf34c715ca
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/SpectorTeleportation.java
@@ -0,0 +1,11 @@
+package one.tranic.sewlia.config.mod.fix;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class SpectorTeleportation {
+    @Comments({"The teleportation of spector players would call absMoveTo directly.",
+            "And when the camera teleported to another region,this would call absMoveTo",
+            "to let the spector player move to another region without any checks, which ",
+            "would trigger the async catcher and crash the server"})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/TaskQueue.java b/src/main/java/one/tranic/sewlia/config/mod/fix/TaskQueue.java
new file mode 100644
index 0000000000000000000000000000000000000000..51c24fa14f56f5f19fa528189a6a01e40b00879f
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/TaskQueue.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.fix;
+
+public class TaskQueue {
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/VillagersDontReleaseMemory.java b/src/main/java/one/tranic/sewlia/config/mod/fix/VillagersDontReleaseMemory.java
new file mode 100644
index 0000000000000000000000000000000000000000..80c80347b9c4a8a00b6c52b92ac0ef5f0e79c4b3
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/VillagersDontReleaseMemory.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.fix;
+
+import one.tranic.sewlia.config.annotation.Comment;
+
+public class VillagersDontReleaseMemory {
+    @Comment("test comment")
+    public static Boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Debug.java b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Debug.java
new file mode 100644
index 0000000000000000000000000000000000000000..90260a1b65a0ca88d84b82ab6a9e92bfe6ae72a8
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Debug.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.fix.anvil_drop;
+
+import one.tranic.sewlia.config.annotation.Comment;
+
+public class Debug {
+    @Comment("Whether to print detailed information to the console when this issue is triggered")
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..8234a65d2aa6b5b3d607cec95453efe41e9f82d9
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/anvil_drop/Enable.java
@@ -0,0 +1,9 @@
+package one.tranic.sewlia.config.mod.fix.anvil_drop;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class Enable {
+    @Comments({"This fix will attempt to prevent the bug in https://github.com/PaperMC/Folia/issues/217, ",
+            "it may cause some unexpected effects but should be harmless."})
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..8fac641a45325f0f5b7c03497e1ce6abed731e94
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/Enable.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.fix.entity_moving;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class Enable {
+    @Comments({"A simple fix of a issue on folia.",
+            "Some times the entity would have a large moment that cross the different tick regions and it would",
+            "make the server crashed)  but sometimes it might doesn't work."})
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/WarnOnDetected.java b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/WarnOnDetected.java
new file mode 100644
index 0000000000000000000000000000000000000000..54d5ac1138baa45f35e6b110112dcae022f9e873
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/fix/entity_moving/WarnOnDetected.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.fix.entity_moving;
+
+public class WarnOnDetected {
+    public static boolean value = true;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..1a6b6d39a0d656beb1d98d6d1f0db713be9a7437
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/Enable.java
@@ -0,0 +1,11 @@
+package one.tranic.sewlia.config.mod.misc.cache_player_profile;
+
+import one.tranic.sewlia.config.annotation.Comments;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Enable {
+    @DisableReload
+    @Comments({"Cache the player profile result on they first join.",
+            "It's useful if Mojang's verification server is down."})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/ResultTimeout.java b/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/ResultTimeout.java
new file mode 100644
index 0000000000000000000000000000000000000000..e3e62aafe376e822295fdb6c9f62847f3aa9ccc9
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/misc/cache_player_profile/ResultTimeout.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.misc.cache_player_profile;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class ResultTimeout {
+    @DisableReload
+    @Comment("The timeout of the cache. Unit: Minutes.")
+    public static int value = 1440;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/Halloween.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/Halloween.java
new file mode 100644
index 0000000000000000000000000000000000000000..5e7a369d27e3319d5dff06214d21509bc68ca7e2
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/Halloween.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.optimization;
+
+public class Halloween {
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/InactiveGoalSelectorThrottle.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/InactiveGoalSelectorThrottle.java
new file mode 100644
index 0000000000000000000000000000000000000000..e657d54881feb75e05438fb7e1c93018f3287ef3
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/InactiveGoalSelectorThrottle.java
@@ -0,0 +1,9 @@
+package one.tranic.sewlia.config.mod.optimization;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class InactiveGoalSelectorThrottle {
+    @Comments({"Throttles the AI goal selector in entity inactive ticks.",
+            "This can improve performance by a few percent, but has minor gameplay implications."})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/RunExtraTasksDuringWaiting.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/RunExtraTasksDuringWaiting.java
new file mode 100644
index 0000000000000000000000000000000000000000..7549383deabee9b461becdd308d9c8708bf4c91d
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/RunExtraTasksDuringWaiting.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.optimization;
+
+public class RunExtraTasksDuringWaiting {
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/Suffocation.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/Suffocation.java
new file mode 100644
index 0000000000000000000000000000000000000000..67a331f19fb7d24e52582fcaed83cf36f17f1995
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/Suffocation.java
@@ -0,0 +1,11 @@
+package one.tranic.sewlia.config.mod.optimization;
+
+import one.tranic.sewlia.config.annotation.Comments;
+
+public class Suffocation {
+    @Comments({"Optimizes the suffocation check by selectively skipping",
+            "the check in a way that still appears vanilla. This should",
+            "be left enabled on most servers, but is provided as a",
+            "configuration option if the vanilla deviation is undesirable."})
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/Enable.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/Enable.java
new file mode 100644
index 0000000000000000000000000000000000000000..78aa064e7829e654117c5ef5078cfedb9ee8fe7c
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/Enable.java
@@ -0,0 +1,8 @@
+package one.tranic.sewlia.config.mod.optimization.projectile_loads;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Enable {
+    @DisableReload
+    public static boolean value = false;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/Max.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/Max.java
new file mode 100644
index 0000000000000000000000000000000000000000..96b15ed1509c8d05f74eb7be025fdd7fe004c0bc
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/Max.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.optimization.projectile_loads;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class Max {
+    @DisableReload
+    @Comment("Controls how many chunks a projectile can load in its lifetime before it gets automatically removed.")
+    public static int value = 10;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/MaxTick.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/MaxTick.java
new file mode 100644
index 0000000000000000000000000000000000000000..dc831edaea729e1909263fbc142cad33eb576f2c
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/projectile_loads/MaxTick.java
@@ -0,0 +1,10 @@
+package one.tranic.sewlia.config.mod.optimization.projectile_loads;
+
+import one.tranic.sewlia.config.annotation.Comment;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+public class MaxTick {
+    @Comment("Controls how many chunks are allowed to be sync loaded by projectiles in a tick.")
+    @DisableReload
+    public static int value = 10;
+}
