From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 26 Oct 2024 13:41:31 +0800
Subject: [PATCH] Fix build


diff --git a/build.gradle.kts b/build.gradle.kts
index 142714c7d47eb5de937ad385c81968148a7e4828..5f5a02743cae8607d3a93f8943cf92b8bb32e4c1 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -25,7 +25,7 @@ abstract class MockitoAgentProvider : CommandLineArgumentProvider {
 // Paper end - configure mockito agent that is needed in newer java versions
 
 dependencies {
-    implementation(project(":folia-api")) // Folia
+    implementation(project(":sewlia-api")) // Folia // Sewlia
     // Paper start
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
