From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 28 Oct 2024 14:20:56 +0800
Subject: [PATCH] Sewlia Config


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 581bd217304e0f9e0b2113c335694805dfb4e2a1..ec4c3c8045488fc8083c77cbf53bd89e7e22dd0c 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -120,6 +120,8 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            one.tranic.sewlia.config.ConfigUtils.reloadConfiguration(false); // Sewlia
+
             io.papermc.paper.plugin.PluginInitializerManager.load(optionset); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
diff --git a/src/main/java/one/tranic/sewlia/config/ConfigScanner.java b/src/main/java/one/tranic/sewlia/config/ConfigScanner.java
new file mode 100644
index 0000000000000000000000000000000000000000..3c70224b62ad935e88f97bdee648fd5da768d648
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/ConfigScanner.java
@@ -0,0 +1,101 @@
+package one.tranic.sewlia.config;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import it.unimi.dsi.fastutil.objects.ObjectList;
+import net.intelie.tinymap.TinyMap;
+import net.intelie.tinymap.TinyMapBuilder;
+import net.intelie.tinymap.TinySet;
+import one.tranic.sewlia.Reflect;
+import one.tranic.sewlia.config.annotation.*;
+import org.jetbrains.annotations.Nullable;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+
+public class ConfigScanner {
+    public static @Nullable TinyMap<Class<?>, String> getClasss() {
+        TinyMapBuilder<Class<?>, String> builder = TinyMap.builder();
+        @Nullable TinySet<Class<?>> classes = Reflect.scanPackage("one.tranic.sewlia.config.mod");
+        if (classes == null || classes.isEmpty()) return null;
+        for (Class<?> clazz : classes) {
+            builder.put(clazz, generateKey(clazz.getName()));
+        }
+        return builder.build();
+    }
+
+    public static String generateKey(String fullClassName) {
+        String cleanedName = fullClassName.replaceFirst("^one\\.tranic\\.sewlia\\.config\\.mod\\.", "");
+        cleanedName = cleanedName.replaceAll("([a-z])([A-Z])", "$1-$2").toLowerCase();
+        cleanedName = cleanedName.replaceAll("_", "-");
+        return cleanedName;
+    }
+
+    public static void processStaticValueFieldWithWrite(Class<?> clazz, String key) {
+        try {
+            Field valueField = clazz.getDeclaredField("value");
+            if (Modifier.isStatic(valueField.getModifiers())) {
+                valueField.setAccessible(true);
+                Object value = valueField.get(null);
+                switch (value) {
+                    case Boolean ignored -> ConfigUtils.getConfiguration().addDefault(key, value);
+                    case String ignored -> ConfigUtils.getConfiguration().addDefault(key, value);
+                    case Integer ignored -> ConfigUtils.getConfiguration().addDefault(key, value);
+                    case Double ignored -> ConfigUtils.getConfiguration().addDefault(key, value);
+                    case List ignored -> ConfigUtils.getConfiguration().addDefault(key, value);
+                    case null, default -> {
+                        return;
+                    }
+                }
+                if (valueField.getAnnotation(Comment.class) != null) {
+                    ConfigUtils.getConfiguration().setComments(key, List.of(valueField.getAnnotation(Comment.class).value()));
+                } else if (valueField.getAnnotation(Comments.class) != null) {
+                    ConfigUtils.getConfiguration().setComments(key, List.of(valueField.getAnnotation(Comments.class).value()));
+                } else if (valueField.getAnnotation(InlineComment.class) != null) {
+                    ConfigUtils.getConfiguration().setInlineComments(key, List.of(valueField.getAnnotation(InlineComment.class).value()));
+                } else if (valueField.getAnnotation(InlineComments.class) != null) {
+                    ConfigUtils.getConfiguration().setInlineComments(key, List.of(valueField.getAnnotation(InlineComments.class).value()));
+                }
+
+                Method wd = clazz.getMethod("WriteDo");
+                if (!Modifier.isStatic(wd.getModifiers())) return;
+                wd.setAccessible(true);
+                wd.invoke(null);
+            }
+        } catch (NoSuchFieldException | NoSuchMethodException | InvocationTargetException | IllegalAccessException ignored) {
+        }
+    }
+
+    public static void processStaticValueFieldWithRead(Class<?> clazz, String key, boolean isReload) {
+        try {
+            Field valueField = clazz.getDeclaredField("value");
+            if (Modifier.isStatic(valueField.getModifiers())) {
+                valueField.setAccessible(true);
+                Object value = valueField.get(null);
+                if (isReload && valueField.getAnnotation(DisableReload.class) != null) return;
+                switch (value) {
+                    case Boolean ignored -> valueField.set(null, ConfigUtils.getConfiguration().getBoolean(key));
+                    case String ignored -> valueField.set(null, ConfigUtils.getConfiguration().getString(key));
+                    case Integer ignored -> valueField.set(null, ConfigUtils.getConfiguration().getInt(key));
+                    case Double ignored -> valueField.set(null, ConfigUtils.getConfiguration().getDouble(key));
+                    case List ignored -> {
+                        List<?> target = ConfigUtils.getConfiguration().getList(key);
+                        if (target == null || target.isEmpty()) return;
+                        valueField.set(null, new ObjectArrayList<>(target));
+                    }
+                    case null, default -> {
+                    }
+                }
+
+                Method dd = clazz.getMethod("ReadDo");
+                if (!Modifier.isStatic(dd.getModifiers())) return;
+                dd.setAccessible(true);
+                dd.invoke(null);
+
+            }
+        } catch (NoSuchFieldException | NoSuchMethodException | InvocationTargetException | IllegalAccessException ignored) {
+        }
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/config/ConfigUtils.java b/src/main/java/one/tranic/sewlia/config/ConfigUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..69aa4868698d11a7921c18c90c4458a3a6b300e8
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/ConfigUtils.java
@@ -0,0 +1,58 @@
+package one.tranic.sewlia.config;
+
+import net.intelie.tinymap.TinyMap;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+
+public class ConfigUtils {
+    private static YamlConfiguration configuration;
+
+    public static YamlConfiguration getConfiguration() {
+        return configuration;
+    }
+
+    public static void reloadConfiguration(boolean isCommandSource) {
+        try {
+            File configFile = new File("sewlia.yaml");
+            if (!configFile.exists()) {
+                configFile.createNewFile();
+            }
+            configuration = YamlConfiguration.loadConfiguration(configFile);
+            if (!isCommandSource) addDefaults(configFile);
+            readAll(isCommandSource);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void addDefaults(@NotNull File file) throws IOException {
+        configuration.options().setHeader(
+                Arrays.asList(
+                        "Sewlia. Configuration",
+                        "Sewlia Github: https://github.com/LevelTranic/Sewlia",
+                        "It is recommended to always keep the latest version, download it at https://tranic.one/downloads/sewlia"
+                )
+        );
+
+        @Nullable TinyMap<Class<?>, String> clz = ConfigScanner.getClasss();
+        if (clz != null && !clz.isEmpty()) {
+            for (Map.Entry<Class<?>, String> entry : clz.entrySet())
+                ConfigScanner.processStaticValueFieldWithWrite(entry.getKey(), entry.getValue());
+        }
+        configuration.options().copyDefaults(true);
+        configuration.save(file);
+    }
+
+    private static void readAll(boolean isReload) {
+        @Nullable TinyMap<Class<?>, String> clz = ConfigScanner.getClasss();
+        if (clz == null || clz.isEmpty()) return;
+        for (Map.Entry<Class<?>, String> entry : clz.entrySet())
+            ConfigScanner.processStaticValueFieldWithRead(entry.getKey(), entry.getValue(), isReload);
+    }
+}
