From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 1 Nov 2024 11:14:17 +0800
Subject: [PATCH] Sewlia Preload Utils


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index ec4c3c8045488fc8083c77cbf53bd89e7e22dd0c..7a4bfd4894a5782695569b2cd51ede88e1d6f603 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -121,6 +121,7 @@ public class Main {
             }
 
             one.tranic.sewlia.config.ConfigUtils.reloadConfiguration(false); // Sewlia
+            new one.tranic.sewlia.preload.Preload(); // Sewlia
 
             io.papermc.paper.plugin.PluginInitializerManager.load(optionset); // Paper
             Bootstrap.bootStrap();
diff --git a/src/main/java/one/tranic/sewlia/preload/Preload.java b/src/main/java/one/tranic/sewlia/preload/Preload.java
new file mode 100644
index 0000000000000000000000000000000000000000..2056586615947c21d9b3b744acbbcc8853a5955f
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/preload/Preload.java
@@ -0,0 +1,25 @@
+package one.tranic.sewlia.preload;
+
+import net.intelie.tinymap.TinySet;
+import one.tranic.sewlia.Reflect;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+public class Preload {
+    private final Logger logger = LoggerFactory.getLogger("PreloadUtils");
+    public Preload() {
+        try {
+            @Nullable TinySet<Preloads> set = Reflect.findAllClass("one.tranic.sewlia.preload.mod", Preloads.class);
+            if (set != null && !set.isEmpty()) {
+                for (Preloads preload : set) {
+                    preload.doPreload();
+                }
+            }
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+        }
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/preload/Preloads.java b/src/main/java/one/tranic/sewlia/preload/Preloads.java
new file mode 100644
index 0000000000000000000000000000000000000000..dd5b919f67d7bfae26efea71d5388ae557177bf9
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/preload/Preloads.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.preload;
+
+public abstract class Preloads {
+    public void doPreload() {}
+}
