From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 3 Nov 2024 11:37:46 +0800
Subject: [PATCH] Add BossBarTask

This patch is inspired by Purpur.

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index e97108dbde4007c881215de8dfee617de6072b6c..8a696841f7df793954d92ede49a6e69ec10b6881 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1236,6 +1236,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.safeShutdown(waitForShutdown, false);
     }
     public void safeShutdown(boolean waitForShutdown, boolean isRestarting) {
+        one.tranic.sewlia.task.BossBarTask.stopAll(); //Sewlia
         this.isRestarting = isRestarting;
         this.hasLoggedStop = true; // Paper - Debugging
         if (isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 0b8f38cad8a5416726bbc2125e6947ac8a582458..0a0dde85ccd4bd58017110c7fe662f1bcc3563a1 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -311,6 +311,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
                 DedicatedServer.LOGGER.info("JMX monitoring enabled");
             }
 
+            one.tranic.sewlia.task.BossBarTask.startAll();
             return true;
         }
     }
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 995ea080be9ee8e3b246723e115e312d2f0984bb..e16cf0ce68e6fb2a0cf0f6f0bc5e8adfcb6b8bd3 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -511,6 +511,7 @@ public abstract class PlayerList {
             scoreboard.addPlayerToTeam(player.getScoreboardName(), collideRuleTeam);
         }
         // Paper end - Configurable player collision
+        one.tranic.sewlia.task.BossBarTask.addToAll(player); // Sewlia
         PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
         // Paper start - Send empty chunk, so players aren't stuck in the world loading screen with our chunk system not sending chunks when dead
         if (player.isDeadOrDying()) {
@@ -632,6 +633,7 @@ public abstract class PlayerList {
     }
     public net.kyori.adventure.text.Component remove(ServerPlayer entityplayer, net.kyori.adventure.text.Component leaveMessage) {
         // Paper end - Fix kick event leave message not being sent
+        one.tranic.sewlia.task.BossBarTask.removeFromAll(entityplayer.getBukkitEntity()); // Sewlia
         ServerLevel worldserver = entityplayer.serverLevel();
 
         entityplayer.awardStat(Stats.LEAVE_GAME);
diff --git a/src/main/java/one/tranic/sewlia/task/BossBarTask.java b/src/main/java/one/tranic/sewlia/task/BossBarTask.java
new file mode 100644
index 0000000000000000000000000000000000000000..cbc4002ff2b865d5c5a528b1d15c4244caa7720c
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/task/BossBarTask.java
@@ -0,0 +1,122 @@
+package one.tranic.sewlia.task;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectIterator;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import net.kyori.adventure.bossbar.BossBar;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+public abstract class BossBarTask implements Runnable {
+    private final Object2ObjectOpenHashMap<UUID, BossBar> listBossbar = new Object2ObjectOpenHashMap<>();
+    private final Logger logger = LoggerFactory.getLogger("BossBarTask");
+    private Thread thread;
+    private boolean running = false;
+
+    abstract BossBar createBossBar();
+
+    abstract void updateBossBar(BossBar bossbar, Player player);
+
+    @Override
+    public void run() {
+        ObjectIterator<Map.Entry<UUID, BossBar>> iter = listBossbar.entrySet().iterator();
+        while (iter.hasNext()) {
+            Map.Entry<UUID, BossBar> entry = iter.next();
+            Player player = Bukkit.getPlayer(entry.getKey());
+            if (player == null) {
+                iter.remove();
+                continue;
+            }
+            updateBossBar(entry.getValue(), player);
+        }
+    }
+
+    private void cancel() {
+        running = false;
+        if (null != this.thread) {
+            int i = 0;
+
+            while (this.thread.isAlive()) {
+                try {
+                    this.thread.join(1000L);
+                    if (++i >= 5) {
+                        logger.warn("Waited {} seconds attempting force stop!", i);
+                    } else if (this.thread.isAlive()) {
+                        logger.warn("Thread {} ({}) failed to exit after {} second(s)", this, this.thread.getState(), i, new Exception("Stack:"));
+                        this.thread.interrupt();
+                    }
+                } catch (InterruptedException var3) {
+                }
+            }
+            this.thread = null;
+        }
+        new ObjectOpenHashSet<>(this.listBossbar.keySet()).forEach(uuid -> {
+            Player player = Bukkit.getPlayer(uuid);
+            if (player != null) removePlayer(player);
+        });
+        this.listBossbar.clear();
+    }
+
+    public boolean removePlayer(Player player) {
+        BossBar bossbar = this.listBossbar.remove(player.getUniqueId());
+        if (bossbar != null) {
+            player.hideBossBar(bossbar);
+            return true;
+        }
+        return false;
+    }
+
+    public void addPlayer(Player player) {
+        removePlayer(player);
+        BossBar bossbar = createBossBar();
+        this.listBossbar.put(player.getUniqueId(), bossbar);
+        this.updateBossBar(bossbar, player);
+        player.showBossBar(bossbar);
+    }
+
+    public boolean hasPlayer(UUID uuid) {
+        return this.listBossbar.containsKey(uuid);
+    }
+
+    public boolean togglePlayer(Player player) {
+        if (removePlayer(player)) return false;
+        addPlayer(player);
+        return true;
+    }
+
+    public void start() {
+        stop();
+        running = true;
+        thread = Thread.ofVirtual().name("Sewlia Bossbar Task Thread").unstarted(() -> {
+            try {
+                while (running) {
+                    this.run();
+                    Thread.sleep(75);
+                }
+            } catch (InterruptedException ignored) {}
+        });
+        thread.start();
+    }
+
+    public void stop() {
+        cancel();
+    }
+
+    public static void startAll() {
+    }
+
+    public static void stopAll() {
+    }
+
+    public static void addToAll(ServerPlayer player) {
+        Player bukkit = player.getBukkitEntity();
+    }
+
+    public static void removeFromAll(Player player) {
+    }
+}
