From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 9 Nov 2024 16:58:59 +0800
Subject: [PATCH] Add RamBarTask

Reference Object: Purpur

diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index fb21572eecb512bdb5ecafcda44f404c2a0e6cde..a5abe913009e0e52e2b4ec3772243630a86f0617 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -304,6 +304,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     public @Nullable String clientBrandName = null; // Paper - Brand support
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
     private boolean tpsBar = false; // Sewlia
+    private boolean ramBar = false; // Sewlia
 
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
@@ -747,6 +748,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         }
 
         if (nbt.contains("Sewlia.TPSBar")) { this.tpsBar = nbt.getBoolean("Sewlia.TPSBar"); } // Sewlia
+        if (nbt.contains("Sewlia.RamBar")) { this.tpsBar = nbt.getBoolean("Sewlia.RamBar"); } // Sewlia
 
     }
 
@@ -825,6 +827,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         }
 
         nbt.putBoolean("Sewlia.TPSBar", this.tpsBar); // Sewlia
+        nbt.putBoolean("Sewlia.RamBar", this.tpsBar); // Sewlia
     }
 
     // CraftBukkit start - World fallback code, either respawn location or global spawn
@@ -3438,4 +3441,12 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     public void tpsBar(boolean tpsBar) {
         this.tpsBar = tpsBar;
     }
+
+    public boolean ramBar() {
+        return this.ramBar;
+    }
+
+    public void ramBar(boolean ramBar) {
+        this.ramBar = ramBar;
+    }
 }
diff --git a/src/main/java/one/tranic/sewlia/command/mod/RamBarCommand.java b/src/main/java/one/tranic/sewlia/command/mod/RamBarCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..5cfdfaeb2dcfbe549337d2110bf41a2433885c47
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/command/mod/RamBarCommand.java
@@ -0,0 +1,67 @@
+package one.tranic.sewlia.command.mod;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
+import one.tranic.sewlia.task.RamBarTask;
+import org.bukkit.Bukkit;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.entity.Player;
+import org.bukkit.util.StringUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class RamBarCommand extends Command {
+    public RamBarCommand() {
+        super("rambar");
+        this.setPermission("sewlia.command.rambar.self");
+    }
+
+    @Override
+    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
+        if (args.length == 0) {
+            if (!(sender instanceof Player)) {
+                sender.sendMessage(Component.text("Only players can use this command", NamedTextColor.RED));
+                return true;
+            }
+            args = new String[]{sender.getName()};
+        } else {
+            if (!sender.hasPermission("sewlia.command.rambar.another")) {
+                sender.sendMessage(Component.text("You do not have permission to switch other people's display!", NamedTextColor.RED));
+                return true;
+            }
+        }
+        for (String target: args) {
+            @Nullable Player player = Bukkit.getPlayer(target);
+            boolean self = (sender instanceof Player) && player != null && player.getUniqueId() == ((Player) sender).getUniqueId();
+            Component output = player != null ? updateTarget(player) : Component.text("Player "+target+" not found");
+            sender.sendMessage(output);
+            if (player != null && !self) player.sendMessage(output);
+        }
+        return true;
+    }
+
+    private Component updateTarget(Player player) {
+        boolean result = RamBarTask.instance().togglePlayer(player);
+        ((CraftPlayer) player).getHandle().ramBar(result);
+
+        return MiniMessage.miniMessage().deserialize("<green>RamBar toggled <onoff> for <target>",
+                Placeholder.component("onoff", Component.translatable(result ? "options.on" : "options.off")
+                        .color(result ? NamedTextColor.GREEN : NamedTextColor.RED)),
+                Placeholder.parsed("target", player.getName()));
+    }
+
+    @Override
+    public @NotNull List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, @NotNull String[] args) {
+        if (args.length < 1 || !sender.hasPermission("sewlia.command.rambar.another")) return List.of();
+        List<String> completion = Bukkit.getOnlinePlayers().stream().map(Player::getName).collect(Collectors.toList());
+        return StringUtil.copyPartialMatches(args[0], completion, new ObjectArrayList<>(completion.size()));
+    }
+}
diff --git a/src/main/java/one/tranic/sewlia/task/BossBarTask.java b/src/main/java/one/tranic/sewlia/task/BossBarTask.java
index 9029450e11bd1436b0b861db0c5279ffa5d26d47..37c0393cdcfba42f6070d19869fce2cd44f00074 100644
--- a/src/main/java/one/tranic/sewlia/task/BossBarTask.java
+++ b/src/main/java/one/tranic/sewlia/task/BossBarTask.java
@@ -109,18 +109,22 @@ public abstract class BossBarTask implements Runnable {
 
     public static void startAll() {
         TPSBarTask.instance().start();
+        RamBarTask.instance().start();
     }
 
     public static void stopAll() {
         TPSBarTask.instance().stop();
+        RamBarTask.instance().stop();
     }
 
     public static void addToAll(ServerPlayer player) {
         Player bukkit = player.getBukkitEntity();
         if (player.tpsBar()) TPSBarTask.instance().addPlayer(bukkit);
+        if (player.ramBar()) RamBarTask.instance().addPlayer(bukkit);
     }
 
     public static void removeFromAll(Player player) {
         TPSBarTask.instance().removePlayer(player);
+        RamBarTask.instance().removePlayer(player);
     }
 }
diff --git a/src/main/java/one/tranic/sewlia/task/RamBarTask.java b/src/main/java/one/tranic/sewlia/task/RamBarTask.java
new file mode 100644
index 0000000000000000000000000000000000000000..5f7e2c384269bc59c2aa155c4a78bcdd2e795f46
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/task/RamBarTask.java
@@ -0,0 +1,124 @@
+package one.tranic.sewlia.task;
+
+import net.kyori.adventure.bossbar.BossBar;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
+import org.bukkit.entity.Player;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryUsage;
+
+public class RamBarTask extends BossBarTask {
+    private static RamBarTask instance;
+    private long allocated = 0L;
+    private long used = 0L;
+    private long xmx = 0L;
+    private long xms = 0L;
+    private float percent = 0F;
+    private int tick = 0;
+    public final String commandRamBarTitle = "<gray>Ram<yellow>:</yellow> <used>/<xmx> (<percent>)";
+    public final BossBar.Overlay commandRamBarProgressOverlay = BossBar.Overlay.NOTCHED_20;
+    public final BossBar.Color commandRamBarProgressColorGood = BossBar.Color.GREEN;
+    public final BossBar.Color commandRamBarProgressColorMedium = BossBar.Color.YELLOW;
+    public final BossBar.Color commandRamBarProgressColorLow = BossBar.Color.RED;
+    public final String commandRamBarTextColorGood = "<gradient:#55ff55:#00aa00><text></gradient>";
+    public final String commandRamBarTextColorMedium = "<gradient:#ffff55:#ffaa00><text></gradient>";
+    public final String commandRamBarTextColorLow = "<gradient:#ff5555:#aa0000><text></gradient>";
+    public final int commandRamBarTickInterval = 20;
+
+    public static RamBarTask instance() {
+        if (instance == null) {
+            instance = new RamBarTask();
+        }
+        return instance;
+    }
+
+    @Override
+    BossBar createBossBar() {
+        return BossBar.bossBar(Component.text(""), 0.0F, instance().getBossBarColor(), commandRamBarProgressOverlay);
+    }
+
+    @Override
+    void updateBossBar(BossBar bossbar, Player player) {
+        bossbar.progress(getBossBarProgress());
+        bossbar.color(getBossBarColor());
+        bossbar.name(MiniMessage.miniMessage().deserialize(commandRamBarTitle,
+                Placeholder.component("allocated", format(this.allocated)),
+                Placeholder.component("used", format(this.used)),
+                Placeholder.component("xmx", format(this.xmx)),
+                Placeholder.component("xms", format(this.xms)),
+                Placeholder.unparsed("percent", ((int) (this.percent * 100)) + "%")
+        ));
+    }
+
+    @Override
+    public void run() {
+        if (++this.tick < commandRamBarTickInterval) {
+            return;
+        }
+        this.tick = 0;
+
+        MemoryUsage heap = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
+        this.allocated = heap.getCommitted();
+        this.used = heap.getUsed();
+        this.xmx = heap.getMax();
+        this.xms = heap.getInit();
+        this.percent = Math.max(Math.min((float) this.used / this.xmx, 1.0F), 0.0F);
+
+        super.run();
+    }
+
+    private float getBossBarProgress() {
+        return this.percent;
+    }
+
+    private BossBar.Color getBossBarColor() {
+        if (this.percent < 0.5F) {
+            return commandRamBarProgressColorGood;
+        } else if (this.percent < 0.75F) {
+            return commandRamBarProgressColorMedium;
+        } else {
+            return commandRamBarProgressColorLow;
+        }
+    }
+
+    public Component format(long v) {
+        String color;
+        if (this.percent < 0.60F) {
+            color = commandRamBarTextColorGood;
+        } else if (this.percent < 0.85F) {
+            color = commandRamBarTextColorMedium;
+        } else {
+            color = commandRamBarTextColorLow;
+        }
+        String value;
+        if (v < 1024) {
+            value = v + "B";
+        } else {
+            int z = (63 - Long.numberOfLeadingZeros(v)) / 10;
+            value = String.format("%.1f%s", (double) v / (1L << (z * 10)), "BKMGTPE".charAt(z));
+        }
+        return MiniMessage.miniMessage().deserialize(color, Placeholder.unparsed("text", value));
+    }
+
+    public long getAllocated() {
+        return this.allocated;
+    }
+
+    public long getUsed() {
+        return this.used;
+    }
+
+    public long getXmx() {
+        return this.xmx;
+    }
+
+    public long getXms() {
+        return this.xms;
+    }
+
+    public float getPercent() {
+        return this.percent;
+    }
+}
\ No newline at end of file
