From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 11 Nov 2024 15:40:34 +0800
Subject: [PATCH] Small modifications in some ServerStatsCounter


diff --git a/src/main/java/net/minecraft/stats/ServerStatsCounter.java b/src/main/java/net/minecraft/stats/ServerStatsCounter.java
index fb7342f7a5008a283c3400c6313c637de8210dfa..0b4c6f5194b337e1dbef5410f5d716373a496020 100644
--- a/src/main/java/net/minecraft/stats/ServerStatsCounter.java
+++ b/src/main/java/net/minecraft/stats/ServerStatsCounter.java
@@ -43,7 +43,7 @@ public class ServerStatsCounter extends StatsCounter {
     private static final Logger LOGGER = LogUtils.getLogger();
     private final MinecraftServer server;
     private final File file;
-    private final Set<Stat<?>> dirty = Sets.newHashSet();
+    private final it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<Stat<?>> dirty = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
 
     public ServerStatsCounter(MinecraftServer server, File file) {
         this.server = server;
@@ -205,7 +205,7 @@ public class ServerStatsCounter extends StatsCounter {
     }
 
     protected String toJson() {
-        Map<StatType<?>, JsonObject> map = Maps.newHashMap();
+        Map<StatType<?>, JsonObject> map = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
         ObjectIterator objectiterator = this.stats.object2IntEntrySet().iterator();
 
         while (objectiterator.hasNext()) {
diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index 5c80e89034c1b3149729684ba4dd4ae26a4261c1..21a7fdc783a00773e14527924a6f8979b0b2b34e 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -246,7 +246,7 @@ public class SpigotConfig
     }
 
     public static boolean disableStatSaving;
-    public static Map<ResourceLocation, Integer> forcedStats = new HashMap<>();
+    public static Map<ResourceLocation, Integer> forcedStats = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
     private static void stats()
     {
         SpigotConfig.disableStatSaving = SpigotConfig.getBoolean( "stats.disable-saving", false );
