From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 12 Nov 2024 15:09:04 +0800
Subject: [PATCH] Small modifications in some CompoundContainer


diff --git a/src/main/java/net/minecraft/world/CompoundContainer.java b/src/main/java/net/minecraft/world/CompoundContainer.java
index 241fec02e6869c638d3a160819b32173a081467b..cb8cb25977e24a1735eebe8b5620d6ccabc8d8a1 100644
--- a/src/main/java/net/minecraft/world/CompoundContainer.java
+++ b/src/main/java/net/minecraft/world/CompoundContainer.java
@@ -4,7 +4,7 @@ import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
 // CraftBukkit start
-import java.util.ArrayList;
+//import java.util.ArrayList;
 import java.util.List;
 import org.bukkit.Location;
 
@@ -18,10 +18,10 @@ public class CompoundContainer implements Container {
     public final Container container2;
 
     // CraftBukkit start - add fields and methods
-    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    public List<HumanEntity> transaction = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
 
     public List<ItemStack> getContents() {
-        List<ItemStack> result = new ArrayList<ItemStack>(this.getContainerSize());
+        List<ItemStack> result = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(this.getContainerSize());
         for (int i = 0; i < this.getContainerSize(); i++) {
             result.add(this.getItem(i));
         }
