From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 2 Nov 2024 23:38:08 +0800
Subject: [PATCH] Luminol: Try fixing folia spector teleportation


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index ade313e0a24d4db662fe3bd05a042eb7682a623f..a9ce8e01ad99b4766ef43672daef6645ac6efa47 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1171,9 +1171,24 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         }
 
         Entity entity = this.getCamera();
+        //Luminol start - Fix folia spector teleportation
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity) && one.tranic.sewlia.config.mod.fix.SpectorTeleportation.value){
+            this.setCamera(this);
+        }
+        //Luminol end
 
         if (entity != this) {
-            if (entity.isAlive()) {
+            // Luminol start - Fix folia spector teleportation
+            var flag = false;
+            var cameraPos = entity.blockPosition();
+            var cameraLevel = entity.level();
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(cameraLevel, cameraPos) && one.tranic.sewlia.config.mod.fix.SpectorTeleportation.value){
+                this.setCamera(this);
+                flag = true;
+            }
+            // Luminol end
+
+            if (entity.isAlive() && !flag) { // Luminol - Fix folia spector teleportation
                 this.absMoveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
                 this.serverLevel().getChunkSource().move(this);
                 if (this.wantsToStopRiding()) {
