From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 2 Nov 2024 12:04:26 +0800
Subject: [PATCH] Leaves: Fix falling block's block location


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 2b997edaab11c2fab76684f0e9208e4b6955b43b..984aed9dbd0d4a8a23a41599983fadd3c6934e66 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -5520,6 +5520,14 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             int j = Mth.floor(y);
             int k = Mth.floor(z);
 
+            // Leaves start - fix falling block location
+            if (this instanceof net.minecraft.world.entity.item.FallingBlockEntity) {
+                if (y < 0.0 && y + 1e-10 > 0.0) {
+                    j = 0;
+                }
+            }
+            // Leaves end - fix falling block location
+
             if (i != this.blockPosition.getX() || j != this.blockPosition.getY() || k != this.blockPosition.getZ()) {
                 this.blockPosition = new BlockPos(i, j, k);
                 this.inBlockState = null;
