From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 2 Nov 2024 12:04:26 +0800
Subject: [PATCH] Leaves: Fix falling block's block location


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index fc6420fab982246a96fcb8843f3317a0f24ffede..8c95b51e6b113f32315fa67bba9de43f6ff54750 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -5893,6 +5893,14 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             int j = Mth.floor(y);
             int k = Mth.floor(z);
 
+            // Leaves start - fix falling block location
+            if (this instanceof net.minecraft.world.entity.item.FallingBlockEntity) {
+                if (y < 0.0 && y + 1e-10 > 0.0) {
+                    j = 0;
+                }
+            }
+            // Leaves end - fix falling block location
+
             if (i != this.blockPosition.getX() || j != this.blockPosition.getY() || k != this.blockPosition.getZ()) {
                 this.blockPosition = new BlockPos(i, j, k);
                 this.inBlockState = null;
