From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Thu, 2 Jun 2022 19:54:09 -0500
Subject: [PATCH] Pufferfish: Entity TTL


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 3cd429e824926c4052e12f33058faa036d14bafb..ca22c0235b81f38a9c2a4ad800e4385530879144 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -180,7 +180,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
-        one.tranic.sewlia.config.mod.optimization.dab.BlacklistedEntities.Do(); // Sewlia - Delayed boot
+        one.tranic.sewlia.config.mod.optimization.dab.BlacklistedEntities.Do(); one.tranic.sewlia.config.mod.optimization.entity_ttl.timeouts.Snowball.Do(); // Sewlia - Delayed boot
         one.tranic.sewlia.command.CommandUtil.register(); // Sewlia - commands
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 19c824b21a8f6ce445cb93308030eb9e47789247..8c453516ed6b7225a0733ac3254247e0a4ee5a7a 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -860,6 +860,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public void tick() {
+        // Pufferfish start - entity TTL
+        if (one.tranic.sewlia.config.mod.optimization.entity_ttl.Enable.value && type != EntityType.PLAYER && type.ttl >= 0 && this.tickCount >= type.ttl) {
+            discard();
+            return;
+        }
+        // Pufferfish end - entity TTL
         this.baseTick();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index c578639784f299a684b613836f0105f41639c334..d0f66ebd05f66f63a40072eda47f1d2a923f80f1 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -317,6 +317,7 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private final int clientTrackingRange;
     private final int updateInterval;
     public boolean dabEnabled = false; // Pufferfish
+    public int ttl = -1; // Pufferfish
     @Nullable
     private String descriptionId;
     @Nullable
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/entity_ttl/timeouts/LlamaSpit.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/entity_ttl/timeouts/LlamaSpit.java
new file mode 100644
index 0000000000000000000000000000000000000000..c71a6d2be52f2135619406d8bb6fbb3b9ac633d6
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/entity_ttl/timeouts/LlamaSpit.java
@@ -0,0 +1,5 @@
+package one.tranic.sewlia.config.mod.optimization.entity_ttl.timeouts;
+
+public class LlamaSpit {
+    public static int value = -1;
+}
diff --git a/src/main/java/one/tranic/sewlia/config/mod/optimization/entity_ttl/timeouts/Snowball.java b/src/main/java/one/tranic/sewlia/config/mod/optimization/entity_ttl/timeouts/Snowball.java
new file mode 100644
index 0000000000000000000000000000000000000000..f96d225fb719e532732423aff20ce3ce693dc0bc
--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/optimization/entity_ttl/timeouts/Snowball.java
@@ -0,0 +1,18 @@
+package one.tranic.sewlia.config.mod.optimization.entity_ttl.timeouts;
+
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.world.entity.EntityType;
+import one.tranic.sewlia.config.ConfigUtils;
+
+import java.util.Locale;
+
+public class Snowball {
+    public static int value = -1;
+
+    public static void Do() {
+        for (EntityType<?> entityType : BuiltInRegistries.ENTITY_TYPE) {
+            String type = EntityType.getKey(entityType).getPath().toLowerCase(Locale.ROOT);
+            entityType.ttl = ConfigUtils.getConfiguration().getInt("optimization.entity-ttl.timeouts." + type, -1);
+        }
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/legacy/FieldRename.java b/src/main/java/org/bukkit/craftbukkit/legacy/FieldRename.java
index 271aad69af4db015970aad842a7bb34dcb6bfd0e..c509ed5914e0f7cd0d8f6bb85680882771402ea2 100644
--- a/src/main/java/org/bukkit/craftbukkit/legacy/FieldRename.java
+++ b/src/main/java/org/bukkit/craftbukkit/legacy/FieldRename.java
@@ -317,7 +317,7 @@ public class FieldRename {
             .change("TOWN_AURA", "MYCELIUM")
             .change("ENCHANTMENT_TABLE", "ENCHANT")
             .change("REDSTONE", "DUST")
-            .change("SNOWBALL", "ITEM_SNOWBALL")
+            .change("Snowball", "ITEM_SNOWBALL")
             .change("SNOW_SHOVEL", "ITEM_SNOWBALL")
             .change("SLIME", "ITEM_SLIME")
             .change("ITEM_CRACK", "ITEM")
