--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/paper/Worker.java
@@ -1,0 +_,17 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.paper;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class Worker {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ExecutorService getDirectExecutorService() {
+        return value ? Executors.newVirtualThreadPerTaskExecutor() : MoreExecutors.newDirectExecutorService();
+    }
+}
