--- /dev/null
+++ b/src/main/java/one/tranic/sewlia/config/mod/feature/virtual_thread/minecraft/io/DimensionData.java
@@ -1,0 +_,14 @@
+package one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io;
+
+import one.tranic.sewlia.config.annotation.DisableReload;
+
+import java.util.concurrent.ThreadFactory;
+
+public class DimensionData {
+    @DisableReload
+    public static boolean value = false;
+
+    public static ThreadFactory getThreadFactory() {
+        return value ? Thread.ofVirtual().factory() : Thread.ofPlatform().factory();
+    }
+}
